---
title: "aldex2_glmm_RNAseq"
author: "Chiamh"
date: "2025-02-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load libraries

```{r}
library(tidyverse)

library("clusterProfiler") #v4.12.6
library("GSEABase") # v1.66.0

library(ALDEx2) # v 1.36.0
library(propr) #version 2.1.2

library(Maaslin2) #v1.18.0

```

Custom functions

```{r}

#Changing df contains the changing genes (up or downregulated)
#background_df contains all detected genes in the l2FC results
run_clusterprofiler <- function(changing_df, background_df, term2gene, term2name=KEGG_term2name){
  
  #the "universe"/background of genes, is all the detected genes in the DEseq2 results and NOT all possible eggnog entries in the bacterial pangenome
  term2gene_filt <- term2gene %>% dplyr::filter(to %in% background_df$feature)
  
  #We actually want to define the universe as all detected genes, which is bigger than the intersect of the background_df and term2gene (previous line)
  other_background <- background_df %>% dplyr::filter(!feature %in% term2gene_filt$to)
  
  if (nrow(other_background) > 0){
    
    other_background_term2gene <- data.frame(from="map00000", to=other_background$feature)
    
    term2gene_universe <- rbind(term2gene_filt, other_background_term2gene)
    
  } else {
    
    term2gene_universe <- term2gene_filt
    
  }
  
  #Add dummy term to term2name
  term2name_dummy <- data.frame(from="map00000", to="No_annotation")
  term2name <- rbind(term2name, term2name_dummy)
  
  #changing_df contains the genes that are significantly different between conditions
  
  #clusterProfiler GSEA or ORA analysis requires a ranked gene list, which contains three features
  #numeric vector: fold change or other type of numerical variable
  #named vector: every number has a name, the corresponding gene ID
  #sorted vector: number should be sorted in decreasing order
  
  #prepare the numeric vector of l2fc of significant DE genes
  geneList <- changing_df$log2FoldChange
  
  #Name the vectors
  names(geneList) <- as.character(changing_df$feature)
  
  #sort the vectors in decreasing order.
  geneList <- sort(geneList, decreasing = TRUE)
  
  #run the ORA analysis
  
  output <- enricher(gene=names(geneList),
                     pvalueCutoff = 0.05, 
                     pAdjustMethod = "BH",
                     minGSSize = 10,
                     maxGSSize = 500,
                     TERM2GENE = term2gene_universe,
                     TERM2NAME = term2name)
  
  
  return(output)
  
}


## Wrapper for GSEA analysis
## fungal and bacteria functional enrichment considered separately (this makes more biological sense)
tally_features_and_enrichments_maaslin2 <- function(site_comparison, 
                                           term2gene_arg, term2name_arg){
  
  l2fc_df <- read_tsv(file=paste0("../data/aldex2_glmm_out/aldex2_gene_clr_",site_comparison,"_l2fc.tsv"), 
                      show_col_types = FALSE) %>%
    dplyr::rename("log2FoldChange"="coef",
                  "pvalue"="pval",
                  "padj"="qval") %>% 
    dplyr::select(feature, log2FoldChange, pvalue, padj, source)
  
  l2fc_df$DE_comparison <- site_comparison
  
  
  bact_l2fc <- l2fc_df %>% dplyr::filter(source=="bacteria")
  fungi_l2fc <- l2fc_df %>% dplyr::filter(source=="fungi")
  
  
  bact_GSEA_results <- GSEA_run(genelist=bact_l2fc$log2FoldChange, 
                                gene_names=bact_l2fc$feature, 
                                term2gene = term2gene_arg, term2name = term2name_arg)
    
  fungi_GSEA_results <- GSEA_run(genelist=fungi_l2fc$log2FoldChange,
                                 gene_names=fungi_l2fc$feature, 
                                 term2gene = term2gene_arg, 
                                 term2name = term2name_arg)
    

  output <- tibble::lst(l2fc_df, 
                        bact_GSEA_results,
                        fungi_GSEA_results)
  
  return(output)
  
}
  
  
GSEA_run <- function(genelist, gene_names, term2gene, term2name = KEGG_term2name){
  
  names(genelist) <- as.character(gene_names)
  # omit any NA values 
  genelist<-na.omit(genelist)
  #sort the genelist in decreasing order.
  genelist <- sort(genelist, decreasing = TRUE)
  
  set.seed(123)
  GSEA_out <- GSEA(geneList = genelist, TERM2GENE = term2gene, 
                   TERM2NAME = term2name, eps= 0, nPermSimple=10000, seed=TRUE)
  
  return(GSEA_out)
  
}


###Already present in clusterprofiler_wrappers.R

#We want to find a set of GO slim metagenome terms that cover all/as much data as possible and represent them in the NES plot
#adapted from https://support.bioconductor.org/p/128407/
#go_ids is a vector of GO terms of interest
#Mode can be "MF" or "BP"
associate_GO_to_slim <- function(go_ids, 
                                 obo_path="../metadata/goslim_metagenomics.obo",
                                 mode="MF"){
  
  myCollection <- GOCollection(go_ids)
  
  # Retrieve GOslims from GO OBO file set
  slim <- getOBOCollection(obo_path)
  
  
  if (mode=="BP"){
    # Retrieve Biological Process (BP) GOslims
    slimdf <- goSlim(myCollection, slim, "BP", verbose)
    # List of GOslims and all GO IDs from `go_ids`
    gomap <- as.list(GOBPOFFSPRING[rownames(slimdf)])
    
  } else if (mode=="MF"){
    # Retrieve Biological Process (MF) GOslims
    slimdf <- goSlim(myCollection, slim, "MF", verbose)
    # List of GOslims and all GO IDs from `go_ids`
    gomap <- as.list(GOMFOFFSPRING[rownames(slimdf)])
  }
  
  
  # Maps `go_ids` to matching GOslims: lapply(gomap, intersect, ids(myCollection))
  
  # Append all mapped GO IDs to `slimdf`
  # `sapply` needed to apply paste() to create semi-colon delimited values
  slimdf$ids <- sapply(lapply(gomap, intersect, ids(myCollection)), paste, collapse=";")
  
  # Remove "character(0) string from "ids" column
  slimdf$ids[slimdf$ids == "character(0)"] <- ""
  
  slimdf[slimdf==""]<-NA
  
  slimdf$GO_slim_parent <- rownames(slimdf)
  
  slimdf$annotation_type <- mode
  
  # Add self-matching GOIDs to "ids" column, if not present
  #i.e the slimmed parent can be itself if it is a slimmed term also
  self_matching_parent_ids  <- intersect(go_ids,slimdf$GO_slim_parent)
  
  slimdf_nonself_subset <- slimdf %>% dplyr::filter(!GO_slim_parent %in% self_matching_parent_ids)
  slimdf_self_subset <- slimdf %>% dplyr::filter(GO_slim_parent %in% self_matching_parent_ids)
  
  
  
  slimdf_self_subset$ids <- ifelse(slimdf_self_subset$Count<=1, 
                                   slimdf_self_subset$GO_slim_parent,
                                   paste0(slimdf_self_subset$ids,";",slimdf_self_subset$GO_slim_parent))
  
  slimdf_self_subset$Count <- str_count(slimdf_self_subset$ids, "GO")
  
  
  slimdf_out <- rbind(slimdf_nonself_subset, slimdf_self_subset)
  slimdf_out <- slimdf_out %>% dplyr::rename(ID=ids) %>% dplyr::select(-Percent)
  
  return(slimdf_out)
  
}

#Function to color code NES bar plots by GO slim (parent categories)
#chosen_parent_ids can be a vector like c("transport","generation of precursor metabolites and energy"...)

prepare_NES_plot_input <- function(df, slimdf, chosen_parent_ids){
  
  slimdf_subset <- slimdf %>% tidyr::separate_rows(ID, sep = ";") %>% 
    dplyr::select(c("ID","Term")) %>% 
    dplyr::filter(Term %in% chosen_parent_ids) %>%
    dplyr::rename(GO_slim_parent_term=Term)
  
  
  df_out <- merge(df, slimdf_subset, by="ID", all.x=TRUE)
  df_out$GO_slim_parent_term <- as.factor(df_out$GO_slim_parent_term)
  
  return(df_out)
  
}

c24 <- c(
  "dodgerblue2", "#E31A1C", # red
  "green4",
  "#6A3D9A", # purple
  "#FF7F00", # orange
  "black", "gold1",
  "skyblue2",
  "palegreen2",
  "#CAB2D6", # lt purple
  "#FDBF6F", # lt orange
  "gray70", "khaki2",
  "maroon", "orchid1", "deeppink1", "blue1", "steelblue4",
  "darkturquoise", "green1", "yellow4", "yellow3",
  "darkorange4", "brown"
)

##Custom functions for aldex2

get_aldex_clr_results <- function(counts_in,seed=123, sample_ids){
  
  set.seed(seed)
  
  #raw clr values
  aldex_clr_results <- aldex.clr(counts_in, conds=sample_ids, mc.samples=1000, verbose = FALSE)
  
  aldex_propr_object <- aldex2propr(aldex_clr_results)
  
  #contains the log-ratio transformed counts as averaged across all Monte Carlo instances
  aldex_avg_clr <- aldex_propr_object@logratio 
  
  output <- tibble::lst(aldex_avg_clr, aldex_clr_results)
  
  return(output)
}


#Custom function for processing maaslin2 results

#assumes maaslin2 default column names
subset_and_recalc_maaslin2 <- function(input_df, feature_of_interest){
  
  df_subset <- input_df %>% dplyr::filter(metadata==feature_of_interest)
  
  df_subset$qval <- p.adjust(df_subset$pval, method="fdr")
  
  return(df_subset)
  
}


```


Load metadata.

```{r}

metadata <- read_tsv("../metadata/skin_mtx_metadata_fmt.txt", show_col_types = FALSE) 

mgx_stats <- read_tsv("../metadata/MGX_QC_stats.txt", show_col_types = FALSE) 

mtx_stats <- read_tsv("../metadata/MTX_QC_stats.txt", show_col_types = FALSE) 

mgx_stats <- merge(mgx_stats, metadata, by = "LIBID", all.x=TRUE)

mtx_stats <- merge(mtx_stats, metadata, by ="LIBID", all.x=TRUE)

#mtx_to_pull and mgx_to_pull refer to the 102 Libraries with paired MTX and MGX data:

mtx_to_pull <- read_tsv("../metadata/mtx_to_pull", col_names = FALSE, show_col_types = FALSE) %>% pull(X1)

mgx_to_pull <- read_tsv("../metadata/mgx_to_pull", col_names = FALSE, show_col_types = FALSE) %>% pull(X1)

mtx_stats_chosen <- mtx_stats %>% dplyr::filter(LIBID %in% mtx_to_pull)

mgx_stats_chosen <- mgx_stats %>% dplyr::filter(LIBID %in% mgx_to_pull)

mtx_mgx_stats_chosen <- read_tsv(file="../metadata/mtx_mgx_stats_chosen.tsv", show_col_types = FALSE)


##########
#MTX Library IDs per site

mtx_Sc_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Sc") %>% pull(mtx_LIBID)
mtx_Ch_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Ch") %>% pull(mtx_LIBID)
mtx_Ac_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Ac") %>% pull(mtx_LIBID)
mtx_Vf_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Vf") %>% pull(mtx_LIBID)
mtx_Tw_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Tw") %>% pull(mtx_LIBID)

#MGX Library IDs per site

mgx_Sc_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Sc") %>% pull(mgx_LIBID)
mgx_Ch_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Ch") %>% pull(mgx_LIBID)
mgx_Ac_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Ac") %>% pull(mgx_LIBID)
mgx_Vf_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Vf") %>% pull(mgx_LIBID)
mgx_Tw_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Tw") %>% pull(mgx_LIBID)


##########
#Load kraken2 filtered and renormalized dataframes.

#MTX data
rna_k2_minimizer_renorm <- lapply(mtx_to_pull, function(x){
  
  df_out <- read_tsv(file=paste0("../data/k2_species_renorm/RNA/",x,"_k2_renorm.s.tsv"),
           show_col_types = FALSE)
  
  return(df_out)
  
})

names(rna_k2_minimizer_renorm) <- mtx_to_pull


#MGX data
k2_minimizer_renorm <- lapply(mgx_to_pull, function(x){
  
  df_out <- read_tsv(file=paste0("../data/k2_species_renorm/DNA/",x,"_k2_renorm.s.tsv"),
           show_col_types = FALSE)
  
  return(df_out)
  
})

names(k2_minimizer_renorm) <- mgx_to_pull

#Load metadata for microbial orthologous groups

bact_og_metadata <- read_tsv("../metadata/bacteria_e5.og_annotations.tsv",
                             col_names = c("bacteria_OG", 
                                           "bacteria_eggnog_cat",
                                           "bacteria_eggnog_desc"),
                             show_col_types = FALSE)

fungi_og_metadata <- read_tsv("../metadata/fungi_e5.og_annotations.tsv",
                             col_names = c("fungi_OG", 
                                           "fungi_eggnog_cat",
                                           "fungi_eggnog_desc"),
                             show_col_types = FALSE)


virus_og_metadata <- read_tsv("../metadata/virus_e5.og_annotations.tsv",
                              col_names = c("virus_OG",
                                            "virus_eggnog_cat",
                                            "virus_eggnog_desc"),
                              show_col_types = FALSE)

###
og_metadata <- rbind(bact_og_metadata %>% dplyr::rename(feature=bacteria_OG,
                                                        eggnog_cat=bacteria_eggnog_cat,
                                                        eggnog_desc=bacteria_eggnog_desc), 
                     fungi_og_metadata %>% dplyr::rename(feature=fungi_OG,
                                                         eggnog_cat=fungi_eggnog_cat,
                                                         eggnog_desc=fungi_eggnog_desc))

og_metadata <- rbind(og_metadata, virus_og_metadata %>% dplyr::rename(feature=virus_OG,
                                                                      eggnog_cat=virus_eggnog_cat,
                                                                      eggnog_desc=virus_eggnog_desc))



# Load GO annotations
GO_annot_all <- read_tsv("../metadata/all_GO_annotations.tsv", show_col_types = FALSE)

#WARNING: This file is large and needs to be copied from figshare into the metadata folder of this repo.
#figshare link: https://figshare.com/articles/dataset/EggNOG_to_Gene_Ontology_annotations/25688532?file=45848136
eggnog_to_GO_all_annot  <- read_tsv("../metadata/eggnog_to_GO_all_annot.tsv", show_col_types = FALSE)

## Load eggnog to all GO annotations in eggnog db 5
eggnog_to_go_all_term2gene <- read_tsv("../metadata/eggnog_with_GO_all_term2gene.txt",
                                       col_names = c("OG", "GO_ID"),
                                       show_col_types=FALSE) %>%
  dplyr::rename(from=GO_ID, to=OG) %>% dplyr::select(from,to)

#Load annotations for Gene Ontology biological processes and molecular functions

#GOs for biological process (BP)
GO_BP_term2name <- GO_annot_all %>% 
  dplyr::filter(ontology=="BP") %>% 
  dplyr::select(GO_ID, GO_desc) %>%
  dplyr::rename(from=GO_ID, to=GO_desc)


eggnog_to_GO_BP_annot <- eggnog_to_GO_all_annot %>% dplyr::filter(ontology == "BP") 
  
eggnog_to_GO_BP_term2gene <- eggnog_to_go_all_term2gene %>% dplyr::filter(from %in% eggnog_to_GO_BP_annot$GO_ID)

#GOs for Molecular Function (MF)
GO_MF_term2name <- GO_annot_all %>% 
  dplyr::filter(ontology=="MF") %>% 
  dplyr::select(GO_ID, GO_desc) %>%
  dplyr::rename(from=GO_ID, to=GO_desc)

eggnog_to_GO_MF_annot <- eggnog_to_GO_all_annot %>% dplyr::filter(ontology == "MF")
  
eggnog_to_GO_MF_term2gene <- eggnog_to_go_all_term2gene %>% dplyr::filter(from %in% eggnog_to_GO_MF_annot$GO_ID)


#Load KEGG pathway and signature modules metadata (e.g https://www.genome.jp/brite/ko00002)

KEGG_module_metadata <- read_tsv("../metadata/KEGG_pathway_modules_metadata",
                                 show_col_types = FALSE,
                                 col_names = c("module", "module_desc"))


#Note the technical replicates besides the 3 bact spike ins

tech_reps <- metadata %>% dplyr::filter(comments == "technical replicate analysis" & subj_region != "3 Bac (VV:LM:PS)")

mtx_to_pull_with_reps <- c(mtx_to_pull, tech_reps %>% pull(LIBID))



```


###  Step 1: Summarize read counts for each feature (IHSMGC Pangene or UniRef90 cluster)

```{r message=FALSE}

mtx_bact_counts <- lapply(mtx_to_pull_with_reps, function(x){
  df <- read_tsv(paste0("../data/MTX_feature_summary/",x,"_bact_OG_counts.tsv"),show_col_types = FALSE) %>% 
    dplyr::rename(feature=bacteria_OG)
  
  df$feature <- gsub(pattern="\\|Bacteria", replacement="", x=df$feature)
  
  return(df)
})

names(mtx_bact_counts) <- mtx_to_pull_with_reps

mgx_bact_counts <- lapply(mgx_to_pull, function(x){
  df <- read_tsv(paste0("../data/MGX_feature_summary/",x,"_bact_OG_counts.tsv"),show_col_types = FALSE) %>%
    dplyr::rename(feature=bacteria_OG)
  
  df$feature <- gsub(pattern="\\|Bacteria", replacement="", x=df$feature)
  
  return(df)
})

names(mgx_bact_counts) <- mgx_to_pull

#################
mtx_fungi_counts <- lapply(mtx_to_pull_with_reps, function(x){
  df <- read_tsv(paste0("../data/MTX_feature_summary/",x,"_fungi_OG_counts.tsv"),show_col_types = FALSE) %>% 
    dplyr::rename(feature=fungi_OG)
  
  df$feature <- gsub(pattern="\\|Fungi", replacement="", x=df$feature)
  
  return(df)
  
})

names(mtx_fungi_counts) <- mtx_to_pull_with_reps


mgx_fungi_counts <- lapply(mgx_to_pull, function(x){
  df <- read_tsv(paste0("../data/MGX_feature_summary/",x,"_fungi_OG_counts.tsv"),show_col_types = FALSE) %>%
    dplyr::rename(feature=fungi_OG)
  
  df$feature <- gsub(pattern="\\|Fungi", replacement="", x=df$feature)
  
  return(df)
  
})

names(mgx_fungi_counts) <- mgx_to_pull

#################
mtx_virus_counts <- lapply(mtx_to_pull_with_reps, function(x){
  read_tsv(paste0("../data/MTX_feature_summary/",x,"_virus_OG_counts.tsv"),show_col_types = FALSE) %>% 
    dplyr::rename(feature=virus_OG)
})

names(mtx_virus_counts) <- mtx_to_pull_with_reps


mgx_virus_counts <- lapply(mgx_to_pull, function(x){
  read_tsv(paste0("../data/MGX_feature_summary/",x,"_virus_OG_counts.tsv"),show_col_types = FALSE) %>%
    dplyr::rename(feature=virus_OG)
})

names(mgx_virus_counts) <- mgx_to_pull


```


### Step 2: Obtain input reads and generate aldex2.clr values


aldex.clr takes a "reads" data.frame with each row being a different gene/microbe and each column represents the read count from that library


```{r}

eligible_mtx_metadata <- metadata %>% dplyr::filter(LIBID %in% c(mtx_to_pull, mgx_to_pull))

site_comparisons_aldex2 <- c( "Ch_vs_Vf", "Tw_vs_Vf")



get_aldex2_input <- function(mtx_count_list, mgx_count_list, input_metadata, cond, ref_cond){
  
  libs_to_pull <- input_metadata %>% dplyr::filter(region %in% c(cond, ref_cond)) %>% pull(LIBID)
  
  
  #######################################################
  ## prepare input read count dataframe for aldex.clr#
  #######################################################
  
  #each row is a different gene/microbe and each column represents the read count from that library
  
  #Obtain count data for eligible mtx and mgx
  rawcounts <- c(mtx_count_list[libs_to_pull], mgx_count_list[libs_to_pull] )%>% do.call("rbind",.)
  
  count_matrix <- pivot_wider(rawcounts %>% 
                                dplyr::select(feature, read_count_sum, LIBID), 
                              names_from = LIBID, values_from = read_count_sum) %>%
    as.data.frame()
  
  rownames(count_matrix) <- count_matrix$feature
  
  count_matrix <- count_matrix %>% dplyr::select(-feature) %>% as.matrix()
  
  #This matrix operation is much faster than mutate_all(~replace(.,is.na(.),0)) for tibbles/df
  count_matrix[is.na(count_matrix)] <- 0 
  
  
  #######################################################
  ## prepare input metadata#
  #######################################################
  
  #the row names of metadata should automatically be ordered in a way that matches the input matrix
  selected_metadata <- input_metadata %>% dplyr::filter(region %in% c(cond, ref_cond)) %>% 
    dplyr::select(LIBID, subject, region, mol_type) %>% as.data.frame()
  
  rownames(selected_metadata) <- selected_metadata$LIBID
  
  selected_metadata <- selected_metadata %>% dplyr::select(-LIBID)
  
  selected_metadata$subject <- as.factor(selected_metadata$subject)
  selected_metadata$region <- as.factor(selected_metadata$region)
  selected_metadata$mol_type <- as.factor(selected_metadata$mol_type)
  
  #reorder rows of the metadata to match the count matrix
  idx <- match(colnames(count_matrix), rownames(selected_metadata))
  selected_metadata <- selected_metadata[idx,]
  
  #a vector of LIBIDs
  RNA_libs <- selected_metadata %>% dplyr::filter(mol_type=="RNA") %>% row.names()
  DNA_libs <- selected_metadata %>% dplyr::filter(mol_type=="DNA") %>% row.names()
  
  #factor relevel such that ref_cond and "DNA" are always the reference conditions
  selected_metadata$region <- relevel(selected_metadata$region, ref_cond)
  selected_metadata$mol_type <- relevel(selected_metadata$mol_type, "DNA")
  
 
  ########################################################
  ## Filter low count features and convert to data.frame##
  ########################################################
  
  #Pre-filtering to keep columns (features) with a minimum median read count for both DNA and RNA seq.
  rna_mat <- count_matrix[, colnames(count_matrix) %in% RNA_libs]
  dna_mat <- count_matrix[, colnames(count_matrix) %in% DNA_libs]
  
  keep <- (Biobase::rowMedians(rna_mat) >= 10) & (Biobase::rowMedians(dna_mat) >= 10)
  count_matrix_filtered <- count_matrix[keep,]
  
  count_df <- count_matrix_filtered %>% as.data.frame()
  
  #######################################################
  ##  Outputs
  #######################################################
  

  output <- tibble::lst(count_df, selected_metadata, comparison=c(cond, ref_cond))
  
  return(output)
  
  
}



aldex2_bact_RNA_Ch_vs_Vf <- get_aldex2_input(mtx_count_list = mtx_bact_counts, 
                                        mgx_count_list = mgx_bact_counts, 
                                        input_metadata = eligible_mtx_metadata,
                                        cond="Ch", ref_cond ="Vf")

aldex2_bact_RNA_Tw_vs_Vf <- get_aldex2_input(mtx_count_list = mtx_bact_counts, 
                                        mgx_count_list = mgx_bact_counts, 
                                        input_metadata = eligible_mtx_metadata,
                                        cond="Tw", ref_cond ="Vf")


aldex2_fungi_RNA_Ch_vs_Vf <- get_aldex2_input(mtx_count_list = mtx_fungi_counts, 
                                        mgx_count_list = mgx_fungi_counts, 
                                        input_metadata = eligible_mtx_metadata,
                                        cond="Ch", ref_cond ="Vf")

aldex2_fungi_RNA_Tw_vs_Vf <- get_aldex2_input(mtx_count_list = mtx_fungi_counts, 
                                        mgx_count_list = mgx_fungi_counts, 
                                        input_metadata = eligible_mtx_metadata,
                                        cond="Tw", ref_cond ="Vf")




```

Analyze bacteria and fungi genes together for aldex2 clr.


```{r}
#
aldex_RNA_Ch_vs_Vf_input <- rbind(aldex2_bact_RNA_Ch_vs_Vf$count_df, aldex2_fungi_RNA_Ch_vs_Vf$count_df)

aldex_RNA_Tw_vs_Vf_input <- rbind(aldex2_bact_RNA_Tw_vs_Vf$count_df, aldex2_fungi_RNA_Tw_vs_Vf$count_df)

```

Use 1000 Monte Carlo instances and get the average log-ratio transformed counts across all these instances

```{r eval=FALSE}

aldex2_RNA_Ch_vs_Vf_results <- get_aldex_clr_results(counts_in=aldex_RNA_Ch_vs_Vf_input,
                                                     sample_ids = colnames(aldex_RNA_Ch_vs_Vf_input),
                                                     seed=123)

aldex2_RNA_Tw_vs_Vf_results <- get_aldex_clr_results(counts_in=aldex_RNA_Tw_vs_Vf_input,
                                                     sample_ids = colnames(aldex_RNA_Tw_vs_Vf_input),
                                                     seed=123)

```
```{r eval=FALSE}
saveRDS(aldex2_RNA_Ch_vs_Vf_results, file="../data/aldex2_RNA_Ch_vs_Vf_results.RDS")

saveRDS(aldex2_RNA_Tw_vs_Vf_results, file="../data/aldex2_RNA_Tw_vs_Vf_results.RDS")

aldex2_RNA_Ch_vs_Vf_clr_df <- aldex2_RNA_Ch_vs_Vf_results$aldex_avg_clr

aldex2_RNA_Tw_vs_Vf_clr_df <- aldex2_RNA_Tw_vs_Vf_results$aldex_avg_clr

write_tsv(aldex2_RNA_Ch_vs_Vf_clr_df %>% rownames_to_column(., var="LIBID"), file="../data/aldex2_RNA_Ch_vs_Vf_clr.tsv")

write_tsv(aldex2_RNA_Tw_vs_Vf_clr_df %>% rownames_to_column(., var="LIBID"), file="../data/aldex2_RNA_Tw_vs_Vf_clr.tsv")


```

Load dataframes (clr averaged across 1000 MC instances) from the previous dataframe.

Columns are features.

The rownames are library IDs. 

```{r}

aldex2_RNA_Ch_vs_Vf_clr_df <- read_tsv("../data/aldex2_RNA_Ch_vs_Vf_clr.tsv", show_col_types = FALSE) %>% as.data.frame()
rownames(aldex2_RNA_Ch_vs_Vf_clr_df) <- aldex2_RNA_Ch_vs_Vf_clr_df$LIBID
aldex2_RNA_Ch_vs_Vf_clr_df <- aldex2_RNA_Ch_vs_Vf_clr_df %>% dplyr::select(-LIBID)


aldex2_RNA_Tw_vs_Vf_clr_df <- read_tsv("../data/aldex2_RNA_Tw_vs_Vf_clr.tsv", show_col_types = FALSE) %>% as.data.frame()
rownames(aldex2_RNA_Tw_vs_Vf_clr_df) <- aldex2_RNA_Tw_vs_Vf_clr_df$LIBID
aldex2_RNA_Tw_vs_Vf_clr_df <- aldex2_RNA_Tw_vs_Vf_clr_df %>% dplyr::select(-LIBID)
```

# Prepare metadata for Maaslin2

Set up the metadata for fixed effects, random effects and interaction terms.

https://github.com/biobakery/omnibus-and-maaslin2-rscripts-and-hmp2-data/blob/master/Maaslin2.md#42-interactions

The first factor levels are the "baseline/reference"
https://forum.biobakery.org/t/interaction-term-missing-from-maaslin2-output/6797

https://forum.biobakery.org/t/paired-samples-by-id/1391/4

Note that, the interaction term is a binary variable by definition since it’s a product of two binary variables and therefore can only take two values: 0 and 1.

With this setup, you don’t need to specify reference as by default 0 will be the reference group in each case and the interaction term is exactly what you want to estimate. You will still need the subject as a random effect.

Row names are Library IDs.

```{r}

#For Ch vs Vf
aldex2_Ch_vs_Vf_metadata <- aldex2_bact_RNA_Ch_vs_Vf$selected_metadata
#The reference level is "Vf" and "DNA"
aldex2_Ch_vs_Vf_metadata$Ch_RNA <- (aldex2_Ch_vs_Vf_metadata$region == "Ch") * 
                                   (aldex2_Ch_vs_Vf_metadata$mol_type == "RNA")
aldex2_Ch_vs_Vf_metadata <- aldex2_Ch_vs_Vf_metadata %>% dplyr::select(subject, region, mol_type, Ch_RNA)

#factor relevel such that "VF" and "DNA" are always the reference conditions
aldex2_Ch_vs_Vf_metadata$region <- relevel(aldex2_Ch_vs_Vf_metadata$region, "Vf")
aldex2_Ch_vs_Vf_metadata$mol_type <- relevel(aldex2_Ch_vs_Vf_metadata$mol_type, "DNA")

###############
#For Tw vs Vf
aldex2_Tw_vs_Vf_metadata <- aldex2_bact_RNA_Tw_vs_Vf$selected_metadata
#The reference level is "Tw" and "DNA"
aldex2_Tw_vs_Vf_metadata$Tw_RNA <- (aldex2_Tw_vs_Vf_metadata$region == "Tw") * 
                                   (aldex2_Tw_vs_Vf_metadata$mol_type == "RNA")
aldex2_Tw_vs_Vf_metadata <- aldex2_Tw_vs_Vf_metadata %>% dplyr::select(subject, region, mol_type, Tw_RNA)

#factor relevel such that "Vf" and "DNA" are always the reference conditions
aldex2_Tw_vs_Vf_metadata$region <- relevel(aldex2_Tw_vs_Vf_metadata$region, "Vf")
aldex2_Tw_vs_Vf_metadata$mol_type <- relevel(aldex2_Tw_vs_Vf_metadata$mol_type, "DNA")


```

# Run GLMM using Maaslin2 as a wrapper

https://forum.biobakery.org/t/paired-samples-with-two-groups/4012/4
Note that, you need to subset your results table to the variable of interest (interaction term) and re-calculate the q-values.

Fixed effect = region (skin site), mol_type and the interaction between the two 

Random effect = "subject"

```{r}
#For Ch vs Vf

aldex2_clr_Ch_vs_Vf_glmm_fit <- Maaslin2(
    input_data = aldex2_RNA_Ch_vs_Vf_clr_df ,
    input_metadata = aldex2_Ch_vs_Vf_metadata,
    output="../data/aldex2_glmm_out/RNA_Ch_vs_Vf",
    min_abundance = -10000, #our current input is already clr transformed and abundance filtered, so this is not relevant
    min_prevalence = 0.0, #our current input is already clr transformed and prevalence filtered
    min_variance = 0.0,
    normalization = "NONE",
    transform = "NONE", #Already clr transformmed
    analysis_method = "LM",
    max_significance = 0.25,
    random_effects = c("subject"),
    fixed_effects = c("region", "mol_type", "Ch_RNA"),
    correction = "BH", #equivalent to "fdr"
    standardize = TRUE,
    cores = 2,
    plot_heatmap = FALSE,
    plot_scatter = FALSE,
    save_scatter = FALSE,
    save_models = FALSE,
    reference = NULL #Not necessary because there are no variables with more than two levels in this design.
)

```

```{r}
#For Tw vs Vf

aldex2_clr_Tw_vs_Vf_glmm_fit <- Maaslin2(
    input_data = aldex2_RNA_Tw_vs_Vf_clr_df ,
    input_metadata = aldex2_Tw_vs_Vf_metadata,
    output="../data/aldex2_glmm_out/RNA_Tw_vs_Vf",
    min_abundance = -10000, #our current input is already clr transformed and abundance filtered, so this is not relevant
    min_prevalence = 0.0, #our current input is already clr transformed and prevalence filtered
    min_variance = 0.0,
    normalization = "NONE",
    transform = "NONE", #Already clr transformmed
    analysis_method = "LM",
    max_significance = 0.25,
    random_effects = c("subject"),
    fixed_effects = c("region", "mol_type", "Tw_RNA"),
    correction = "BH", #equivalent to "fdr"
    standardize = TRUE,
    cores = 2,
    plot_heatmap = FALSE,
    plot_scatter = FALSE,
    save_scatter = FALSE,
    save_models = FALSE,
    reference = NULL #Not necessary because there are no variables with more than two levels in this design.
)

```

Load the results.. subset by the correct metadata/coeff "Ch_RNA or Tw_RNA" since that is the interaction term of interest.

Re-calculate the FDR after subsetting for the interaction term of interest

```{r}

#aldex.clr uses the log2 transform
#Maaslin2 adds an "X" in front of any gene feature which starts with a number. E.g.2Z7RS becomes X2Z7RS. Remove this leading X
#Check that 33DPX is not affected

aldex2_clr_Ch_vs_Vf_glmm_results <- read_tsv("../data/aldex2_glmm_out/RNA_Ch_vs_Vf/all_results.tsv", 
                                             show_col_types = FALSE) %>%
                                             subset_and_recalc_maaslin2(input_df=., 
                                                                        feature_of_interest="Ch_RNA")


aldex2_clr_Ch_vs_Vf_glmm_results$feature <- gsub(pattern="^X", replacement="", x=aldex2_clr_Ch_vs_Vf_glmm_results$feature)

#Label fungal vs bacterial gene features in a separate column

aldex2_clr_Ch_vs_Vf_glmm_results$source <- ifelse(aldex2_clr_Ch_vs_Vf_glmm_results$feature %in% rownames(aldex2_bact_RNA_Ch_vs_Vf$count_df),
                                                  "bacteria", "fungi")


write_tsv(aldex2_clr_Ch_vs_Vf_glmm_results, "../data/aldex2_glmm_out/aldex2_gene_clr_Ch_vs_Vf_l2fc.tsv")


aldex2_clr_Tw_vs_Vf_glmm_results <- read_tsv("../data/aldex2_glmm_out/RNA_Tw_vs_Vf/all_results.tsv", 
                                             show_col_types = FALSE) %>%
                                             subset_and_recalc_maaslin2(input_df=., 
                                                                        feature_of_interest="Tw_RNA")

aldex2_clr_Tw_vs_Vf_glmm_results$feature <- gsub(pattern="^X", replacement="", x=aldex2_clr_Tw_vs_Vf_glmm_results$feature)

aldex2_clr_Tw_vs_Vf_glmm_results$source <- ifelse(aldex2_clr_Tw_vs_Vf_glmm_results$feature %in% rownames(aldex2_bact_RNA_Tw_vs_Vf$count_df),
                                                  "bacteria", "fungi")


write_tsv(aldex2_clr_Tw_vs_Vf_glmm_results, "../data/aldex2_glmm_out/aldex2_gene_clr_Tw_vs_Vf_l2fc.tsv")

```

Perform gene set enrichment analysis results on maaslin2 results, based on the coefficient of the interaction term (is a log fold change due to the clr transformation)

Bacterial and Fungal enrichments computed separately.

```{r}
#Molecular function
site_clr_enrichments_GO_MF_GSEA <- lapply(site_comparisons_aldex2, function(x){
  
  list_out <- tally_features_and_enrichments_maaslin2(x, 
                                             term2gene_arg = eggnog_to_GO_MF_term2gene,
                                             term2name_arg = GO_MF_term2name)
  
  return(list_out)
  
})


names(site_clr_enrichments_GO_MF_GSEA) <- site_comparisons_aldex2

saveRDS(site_clr_enrichments_GO_MF_GSEA, file="../data/aldex2_glmm_out/site_clr_enrichments_GO_MF_GSEA.RDS")


#Biological process
site_clr_enrichments_GO_BP_GSEA <- lapply(site_comparisons_aldex2, function(x){
  
  list_out <- tally_features_and_enrichments_maaslin2(x, 
                                             term2gene_arg = eggnog_to_GO_BP_term2gene,
                                             term2name_arg = GO_BP_term2name)
  
  return(list_out)
  
})


names(site_clr_enrichments_GO_BP_GSEA) <- site_comparisons_aldex2

saveRDS(site_clr_enrichments_GO_BP_GSEA, file="../data/aldex2_glmm_out/site_clr_enrichments_GO_BP_GSEA.RDS")


```


## Normalized enrichment score plots for aldex.clr-Maaslin2 results

Comparison: Cheeks vs Vf. Bacteria and Fungi
```{r}


gene_clr_Ch_vs_Vf_bact_GO_MF_GSEA <- site_clr_enrichments_GO_MF_GSEA$Ch_vs_Vf$bact_GSEA_results@result


gene_clr_Ch_vs_Vf_fungi_GO_MF_GSEA <- site_clr_enrichments_GO_MF_GSEA$Ch_vs_Vf$fungi_GSEA_results@result


gene_clr_Ch_vs_Vf_bact_GO_BP_GSEA <- site_clr_enrichments_GO_BP_GSEA$Ch_vs_Vf$bact_GSEA_results@result


gene_clr_Ch_vs_Vf_fungi_GO_BP_GSEA <- site_clr_enrichments_GO_BP_GSEA$Ch_vs_Vf$fungi_GSEA_results@result


#################
###GO Slimming###
#################
#For Molecular function
gene_clr_Ch_vs_Vf_bact_GO_MF_GSEA_slim <- associate_GO_to_slim(gene_clr_Ch_vs_Vf_bact_GO_MF_GSEA$ID, mode="MF")

gene_clr_Ch_vs_Vf_bact_GO_MF_GSEA_fmt <- prepare_NES_plot_input(df=gene_clr_Ch_vs_Vf_bact_GO_MF_GSEA,
                                                               slimdf=gene_clr_Ch_vs_Vf_bact_GO_MF_GSEA_slim,
                                                               chosen_parent_ids = c("nucleic acid binding",
                                                                                     "antioxidant activity"))

gene_clr_Ch_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df <- gene_clr_Ch_vs_Vf_bact_GO_MF_GSEA_fmt %>% dplyr::select(Description, NES, GO_slim_parent_term)

gene_clr_Ch_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df$organism <- "Bacteria"
gene_clr_Ch_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df$GO_level <- "Molecular function"

#For fungi, MF.

gene_clr_Ch_vs_Vf_fungi_GO_MF_GSEA_slim <- associate_GO_to_slim(gene_clr_Ch_vs_Vf_fungi_GO_MF_GSEA$ID, mode="MF")

gene_clr_Ch_vs_Vf_fungi_GO_MF_GSEA_fmt <- prepare_NES_plot_input(df=gene_clr_Ch_vs_Vf_fungi_GO_MF_GSEA,
                                                               slimdf=gene_clr_Ch_vs_Vf_fungi_GO_MF_GSEA_slim,
                                                               chosen_parent_ids = c("nucleic acid binding",
                                                                                     "antioxidant activity"))

gene_clr_Ch_vs_Vf_fungi_GO_MF_GSEA_NES_condensed_df <- gene_clr_Ch_vs_Vf_fungi_GO_MF_GSEA_fmt %>% dplyr::select(Description, NES, GO_slim_parent_term)


gene_clr_Ch_vs_Vf_fungi_GO_MF_GSEA_NES_condensed_df$organism <- "Fungi"
gene_clr_Ch_vs_Vf_fungi_GO_MF_GSEA_NES_condensed_df$GO_level <- "Molecular function"



##############
#For biological process

gene_clr_Ch_vs_Vf_bact_GO_BP_GSEA_slim <- associate_GO_to_slim(gene_clr_Ch_vs_Vf_bact_GO_BP_GSEA$ID, mode="BP")

gene_clr_Ch_vs_Vf_bact_GO_BP_GSEA_fmt <- prepare_NES_plot_input(df=gene_clr_Ch_vs_Vf_bact_GO_BP_GSEA,
                                                               slimdf=gene_clr_Ch_vs_Vf_bact_GO_BP_GSEA_slim,
                                                               chosen_parent_ids = c(
                                                       "transport", "translation", "carbohydrate metabolic process",
                                                       "protein-containing complex assembly",
                                                       "generation of precursor metabolites and energy"))



gene_clr_Ch_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df <- gene_clr_Ch_vs_Vf_bact_GO_BP_GSEA_fmt %>% 
                                                                            dplyr::filter(Description %in% 
                                                                                            c("glucose catabolic process",
                                                                                              "canonical glycolysis",
                                                                                              "proton motive force-driven ATP synthesis",
                                                                                              "NADH regeneration",
                                                                                              "cotranslational protein targeting to membrane",
                                                                                              "ribosomal large subunit assembly",
                                                                                              "ribosomal small subunit assembly")) %>% dplyr::select(Description, NES, GO_slim_parent_term)

gene_clr_Ch_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df$organism <- "Bacteria"
gene_clr_Ch_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df$GO_level <- "Biological process"

#

gene_clr_Ch_vs_Vf_fungi_GO_BP_GSEA_slim <- associate_GO_to_slim(gene_clr_Ch_vs_Vf_fungi_GO_BP_GSEA$ID, mode="BP")

gene_clr_Ch_vs_Vf_fungi_GO_BP_GSEA_fmt <- prepare_NES_plot_input(df=gene_clr_Ch_vs_Vf_fungi_GO_BP_GSEA,
                                                               slimdf=gene_clr_Ch_vs_Vf_fungi_GO_BP_GSEA_slim,
                                                               chosen_parent_ids = c("nitrogen compound metabolic process",
                                                       "RNA metabolic process",
                                                       "DNA metabolic process",
                                                       "response to chemical",
                                                       "metabolic process", "response to stress"))

gene_clr_Ch_vs_Vf_fungi_GO_BP_GSEA_NES_condensed_df <- gene_clr_Ch_vs_Vf_fungi_GO_BP_GSEA_fmt %>%
                                                                              dplyr::filter(Description %in% 
                                                                                              c("protein lipidation",
                                                                                            "lipoprotein biosynthetic process",
                                                                                            "peptide metabolic process")) %>% dplyr::select(Description, NES, GO_slim_parent_term)


gene_clr_Ch_vs_Vf_fungi_GO_BP_GSEA_NES_condensed_df$organism <- "Fungi"
gene_clr_Ch_vs_Vf_fungi_GO_BP_GSEA_NES_condensed_df$GO_level <- "Biological process"


##Plotting
gene_clr_Ch_vs_Vf_all_GO_GSEA_NES_condensed_df <- do.call("rbind", lst(gene_clr_Ch_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df,
                                                              gene_clr_Ch_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df,
                                                              gene_clr_Ch_vs_Vf_fungi_GO_BP_GSEA_NES_condensed_df,
                                                              gene_clr_Ch_vs_Vf_fungi_GO_MF_GSEA_NES_condensed_df))


gene_clr_Ch_vs_Vf_all_GO_GSEA_NES_condensed_df$direction <- ifelse(gene_clr_Ch_vs_Vf_all_GO_GSEA_NES_condensed_df$NES >0,
                                                          "enriched_in_Ch", "enriched_in_Vf")



gene_clr_Ch_vs_Vf_all_GO_GSEA_NES_condensed_plot <- ggplot(gene_clr_Ch_vs_Vf_all_GO_GSEA_NES_condensed_df, aes(x=Description, y=NES, fill = direction)) +
  geom_col(colour="black",position="dodge") +  
  geom_hline(yintercept=0, linetype="longdash") +
  scale_fill_manual(values = c24) +
  theme(axis.text.x = element_text(vjust = 0.5, hjust=1)) +
  coord_flip() +
  labs(y="Normalized Enrichment Score") + 
   facet_grid(rows=vars(organism),cols=vars(GO_level),
             scales = "free", space = "free",
             labeller = label_wrap_gen(width = 1.5, multi_line = TRUE)) + 
  theme_classic() +
  theme(panel.spacing = unit(.05, "lines"),
        panel.border = element_rect(color = "black", fill = NA, size = 0.2), 
       strip.background = element_rect(color = "black", size = 1)) + ggtitle("Cheeks vs. Volar forearms")


gene_clr_Ch_vs_Vf_all_GO_GSEA_NES_condensed_plot

```
Toe webs vs Vf


```{r}

gene_clr_Tw_vs_Vf_bact_GO_MF_GSEA <- site_clr_enrichments_GO_MF_GSEA$Tw_vs_Vf$bact_GSEA_results@result


gene_clr_Tw_vs_Vf_bact_GO_BP_GSEA <- site_clr_enrichments_GO_BP_GSEA$Tw_vs_Vf$bact_GSEA_results@result


gene_clr_Tw_vs_Vf_bact_GO_MF_GSEA_slim <- associate_GO_to_slim(gene_clr_Tw_vs_Vf_bact_GO_MF_GSEA$ID, mode="MF")

gene_clr_Tw_vs_Vf_bact_GO_MF_GSEA_fmt <- prepare_NES_plot_input(df=gene_clr_Tw_vs_Vf_bact_GO_MF_GSEA,
                                                               slimdf=gene_clr_Tw_vs_Vf_bact_GO_MF_GSEA_slim,
                                                               chosen_parent_ids = c("catalytic activity",
                                                                                     "transferase activity",
                                                                                     "hydrolase activity",
                                                                                     "oxidoreductase activity",
                                                                                     "antioxidant activity"))


gene_clr_Tw_vs_Vf_bact_GO_BP_GSEA_slim <- associate_GO_to_slim(gene_clr_Tw_vs_Vf_bact_GO_BP_GSEA$ID, mode="BP")

gene_clr_Tw_vs_Vf_bact_GO_BP_GSEA_fmt <- prepare_NES_plot_input(df=gene_clr_Tw_vs_Vf_bact_GO_BP_GSEA,
                                                               slimdf=gene_clr_Tw_vs_Vf_bact_GO_BP_GSEA_slim,
                                                               chosen_parent_ids = c("nitrogen compound metabolic process",
                                                                                     "RNA metabolic process",
                                                                                     "DNA metabolic process",
                                                                                     "transport",
                                                                                     "small molecule metabolic process",
                                                                                     "generation of precursor metabolites and energy","response to stress","response to abiotic stimulus","response to chemical"))


gene_clr_Tw_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df <- gene_clr_Tw_vs_Vf_bact_GO_BP_GSEA_fmt %>%
                                                                              dplyr::filter(Description %in% 
                                                                                              c("tricarboxylic acid cycle",
"cell cycle DNA replication",
"amino acid metabolic process",
"thiamine biosynthetic process",
"protoporphyrinogen IX metabolic process",
"RNA methylation",
"sodium ion transmembrane transport",
"amino acid transport",
"carboxylic acid transport",
"transposition")) %>% dplyr::select(Description, NES, GO_slim_parent_term)

gene_clr_Tw_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df$organism <- "Bacteria"
gene_clr_Tw_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df$GO_level <- "Biological process"


gene_clr_Tw_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df <- gene_clr_Tw_vs_Vf_bact_GO_MF_GSEA_fmt %>%
                                                                              dplyr::filter(Description %in% 
                                                                                              c("ribonucleotide binding",
                                                                                            "DNA helicase activity",
                                                                                            "exonuclease activity",
                                                                                            "RNA methyltransferase activity","ligase activity, forming carbon-nitrogen bonds",
"hydrolase activity, acting on ester bonds",
"kinase activity",
"antioxidant activity",
"peroxidase activity",
"fibronectin binding")) %>% dplyr::select(Description, NES, GO_slim_parent_term)

gene_clr_Tw_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df$organism <- "Bacteria"
gene_clr_Tw_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df$GO_level <- "Molecular function"


gene_clr_Tw_vs_Vf_all_GO_GSEA_NES_condensed_df <- do.call("rbind", lst(gene_clr_Tw_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df,
                                                              gene_clr_Tw_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df,
                                                           ))


gene_clr_Tw_vs_Vf_all_GO_GSEA_NES_condensed_df$direction <- ifelse(gene_clr_Tw_vs_Vf_all_GO_GSEA_NES_condensed_df$NES >0,
                                                          "enriched_in_Tw", "enriched_in_Vf")




gene_clr_Tw_vs_Vf_all_GO_GSEA_NES_condensed_plot <- ggplot(gene_clr_Tw_vs_Vf_all_GO_GSEA_NES_condensed_df, aes(x=Description, y=NES, fill = direction)) +
  geom_col(colour="black",position="dodge") +  
  geom_hline(yintercept=0, linetype="longdash") +
  scale_fill_manual(values = c24) +
  theme(axis.text.x = element_text(vjust = 0.5, hjust=1)) +
  coord_flip() +
  labs(y="Normalized Enrichment Score") + 
   facet_grid(rows=vars(organism),cols=vars(GO_level),
             scales = "free", space = "free",
             labeller = label_wrap_gen(width = 1.5, multi_line = TRUE)) + 
  theme_classic() +
  theme(panel.spacing = unit(.05, "lines"),
        panel.border = element_rect(color = "black", fill = NA, size = 0.2), 
       strip.background = element_rect(color = "black", size = 1)) + ggtitle("Toe webs vs. Volar forearms")


gene_clr_Tw_vs_Vf_all_GO_GSEA_NES_condensed_plot




```




##Appendix##

bact and fungi were run separately here... Is this the right approach?

need to add conds.. (reflected in the sample_ids actually)

```{r eval=FALSE}

aldex2_bact_RNA_Ch_vs_Vf_results <- get_aldex_clr_results(counts_in=aldex2_bact_RNA_Ch_vs_Vf$count_df,seed=123)

aldex2_bact_RNA_Tw_vs_Vf_results <- get_aldex_clr_results(counts_in=aldex2_bact_RNA_Tw_vs_Vf$count_df,seed=123)

aldex2_fungi_RNA_Ch_vs_Vf_results <- get_aldex_clr_results(counts_in=aldex2_fungi_RNA_Ch_vs_Vf$count_df,seed=123)

aldex2_fungi_RNA_Tw_vs_Vf_results <- get_aldex_clr_results(counts_in=aldex2_fungi_RNA_Tw_vs_Vf$count_df,seed=123)

saveRDS(aldex2_bact_RNA_Ch_vs_Vf_results, file="../data/aldex2_bact_RNA_Ch_vs_Vf_results.RDS")

saveRDS(aldex2_bact_RNA_Tw_vs_Vf_results, file="../data/aldex2_bact_RNA_Tw_vs_Vf_results.RDS")

saveRDS(aldex2_fungi_RNA_Ch_vs_Vf_results, file="../data/aldex2_fungi_RNA_Ch_vs_Vf_results.RDS")

saveRDS(aldex2_fungi_RNA_Tw_vs_Vf_results, file="../data/aldex2_fungi_RNA_Tw_vs_Vf_results.RDS")



aldex2_bact_RNA_Ch_vs_Vf_clr_df <- aldex2_bact_RNA_Ch_vs_Vf_results$aldex_avg_clr
rownames(aldex2_bact_RNA_Ch_vs_Vf_clr_df) <- getSampleIDs(aldex2_bact_RNA_Ch_vs_Vf_results$aldex_clr_results)

aldex2_bact_RNA_Tw_vs_Vf_clr_df <- aldex2_bact_RNA_Tw_vs_Vf_results$aldex_avg_clr
rownames(aldex2_bact_RNA_Tw_vs_Vf_clr_df) <- getSampleIDs(aldex2_bact_RNA_Tw_vs_Vf_results$aldex_clr_results)

aldex2_fungi_RNA_Ch_vs_Vf_clr_df <- aldex2_fungi_RNA_Ch_vs_Vf_results$aldex_avg_clr
rownames(aldex2_fungi_RNA_Ch_vs_Vf_clr_df) <- getSampleIDs(aldex2_fungi_RNA_Ch_vs_Vf_results$aldex_clr_results)

aldex2_fungi_RNA_Tw_vs_Vf_clr_df <- aldex2_fungi_RNA_Tw_vs_Vf_results$aldex_avg_clr
rownames(aldex2_fungi_RNA_Tw_vs_Vf_clr_df) <- getSampleIDs(aldex2_fungi_RNA_Tw_vs_Vf_results$aldex_clr_results)

#might be outdated if we are not going down this route
write_tsv(aldex2_bact_RNA_Ch_vs_Vf_results$aldex_avg_clr %>% rownames_to_column(., var="LIBID"), file="../data/aldex2_bact_RNA_Ch_vs_Vf_clr.tsv")

write_tsv(aldex2_bact_RNA_Tw_vs_Vf_results$aldex_avg_clr %>% rownames_to_column(., var="LIBID"), file="../data/aldex2_bact_RNA_Tw_vs_Vf_clr.tsv")

write_tsv(aldex2_fungi_RNA_Ch_vs_Vf_results$aldex_avg_clr %>% rownames_to_column(., var="LIBID"), file="../data/aldex2_fungi_RNA_Ch_vs_Vf_clr.tsv")

write_tsv(aldex2_fungi_RNA_Tw_vs_Vf_results$aldex_avg_clr %>% rownames_to_column(., var="LIBID"), file="../data/aldex2_fungi_RNA_Tw_vs_Vf_clr.tsv")

```

Load results in the previous code chunk to save time.

```{r eval=FALSE}

aldex2_bact_RNA_Ch_vs_Vf_clr_df <- read_tsv("../data/aldex2_bact_RNA_Ch_vs_Vf_clr.tsv")

aldex2_bact_RNA_Tw_vs_Vf_clr_df <- read_tsv("../data/aldex2_bact_RNA_Tw_vs_Vf_clr.tsv")

aldex2_fungi_RNA_Ch_vs_Vf_clr_df <- read_tsv("../data/aldex2_fungi_RNA_Ch_vs_Vf_clr.tsv")

aldex2_fungi_RNA_Tw_vs_Vf_clr_df <- read_tsv("../data/aldex2_fungi_RNA_Tw_vs_Vf_clr.tsv")

```





