---
title: "OG level DE"
author: "Chiamh"
date: '2024-04-24'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load libraries and custom functions

```{r}
library(tidyverse)
library(reshape2)
library(ggpubr)

library(DESeq2)
library("clusterProfiler") #v4.12.6
library("GSEABase") # v1.66.0

library(vegan)

library(EnhancedVolcano)

#functions
source("custom_functions.R")
source("DESeq2_wrappers.R")
source("clusterprofiler_wrappers.R")

```


Load metadata and supporting information

```{r}

metadata <- read_tsv("../metadata/skin_mtx_metadata_fmt.txt", show_col_types = FALSE) 

mgx_stats <- read_tsv("../metadata/MGX_QC_stats.txt", show_col_types = FALSE) 

mtx_stats <- read_tsv("../metadata/MTX_QC_stats.txt", show_col_types = FALSE) 

mgx_stats <- merge(mgx_stats, metadata, by = "LIBID", all.x=TRUE)

mtx_stats <- merge(mtx_stats, metadata, by ="LIBID", all.x=TRUE)

#mtx_to_pull and mgx_to_pull refer to the 102 Libraries with paired MTX and MGX data:

mtx_to_pull <- read_tsv("../metadata/mtx_to_pull", col_names = FALSE, show_col_types = FALSE) %>% pull(X1)

mgx_to_pull <- read_tsv("../metadata/mgx_to_pull", col_names = FALSE, show_col_types = FALSE) %>% pull(X1)

mtx_stats_chosen <- mtx_stats %>% dplyr::filter(LIBID %in% mtx_to_pull)

mgx_stats_chosen <- mgx_stats %>% dplyr::filter(LIBID %in% mgx_to_pull)

mtx_mgx_stats_chosen <- read_tsv(file="../metadata/mtx_mgx_stats_chosen.tsv", show_col_types = FALSE)


##########
#MTX Library IDs per site

mtx_Sc_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Sc") %>% pull(mtx_LIBID)
mtx_Ch_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Ch") %>% pull(mtx_LIBID)
mtx_Ac_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Ac") %>% pull(mtx_LIBID)
mtx_Vf_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Vf") %>% pull(mtx_LIBID)
mtx_Tw_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Tw") %>% pull(mtx_LIBID)

#MGX Library IDs per site

mgx_Sc_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Sc") %>% pull(mgx_LIBID)
mgx_Ch_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Ch") %>% pull(mgx_LIBID)
mgx_Ac_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Ac") %>% pull(mgx_LIBID)
mgx_Vf_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Vf") %>% pull(mgx_LIBID)
mgx_Tw_ids <- mtx_mgx_stats_chosen %>% dplyr::filter(region=="Tw") %>% pull(mgx_LIBID)

##########
#Load kraken2 filtered and renormalized dataframes.

#MTX data
rna_k2_minimizer_renorm <- lapply(mtx_to_pull, function(x){
  
  df_out <- read_tsv(file=paste0("../data/k2_species_renorm/RNA/",x,"_k2_renorm.s.tsv"),
           show_col_types = FALSE)
  
  return(df_out)
  
})

names(rna_k2_minimizer_renorm) <- mtx_to_pull


#MGX data
k2_minimizer_renorm <- lapply(mgx_to_pull, function(x){
  
  df_out <- read_tsv(file=paste0("../data/k2_species_renorm/DNA/",x,"_k2_renorm.s.tsv"),
           show_col_types = FALSE)
  
  return(df_out)
  
})

names(k2_minimizer_renorm) <- mgx_to_pull

#Load metadata for microbial orthologous groups

bact_og_metadata <- read_tsv("../metadata/bacteria_e5.og_annotations.tsv",
                             col_names = c("bacteria_OG", 
                                           "bacteria_eggnog_cat",
                                           "bacteria_eggnog_desc"),
                             show_col_types = FALSE)

fungi_og_metadata <- read_tsv("../metadata/fungi_e5.og_annotations.tsv",
                             col_names = c("fungi_OG", 
                                           "fungi_eggnog_cat",
                                           "fungi_eggnog_desc"),
                             show_col_types = FALSE)


virus_og_metadata <- read_tsv("../metadata/virus_e5.og_annotations.tsv",
                              col_names = c("virus_OG",
                                            "virus_eggnog_cat",
                                            "virus_eggnog_desc"),
                              show_col_types = FALSE)

###
og_metadata <- rbind(bact_og_metadata %>% dplyr::rename(feature=bacteria_OG,
                                                        eggnog_cat=bacteria_eggnog_cat,
                                                        eggnog_desc=bacteria_eggnog_desc), 
                     fungi_og_metadata %>% dplyr::rename(feature=fungi_OG,
                                                         eggnog_cat=fungi_eggnog_cat,
                                                         eggnog_desc=fungi_eggnog_desc))

og_metadata <- rbind(og_metadata, virus_og_metadata %>% dplyr::rename(feature=virus_OG,
                                                                      eggnog_cat=virus_eggnog_cat,
                                                                      eggnog_desc=virus_eggnog_desc))



# Load GO annotations
GO_annot_all <- read_tsv("../metadata/all_GO_annotations.tsv", show_col_types = FALSE)

#WARNING: This file is large and needs to be copied from figshare into the metadata folder of this repo.
#figshare link: https://figshare.com/articles/dataset/EggNOG_to_Gene_Ontology_annotations/25688532?file=45848136
eggnog_to_GO_all_annot  <- read_tsv("../metadata/eggnog_to_GO_all_annot.tsv", show_col_types = FALSE)
```

```{r}
## Load eggnog to all GO annotations in eggnog db 5
eggnog_to_go_all_term2gene <- read_tsv("../metadata/eggnog_with_GO_all_term2gene.txt",
                                       col_names = c("OG", "GO_ID"),
                                       show_col_types=FALSE) %>%
  dplyr::rename(from=GO_ID, to=OG) %>% dplyr::select(from,to)

#Load annotations for Gene Ontology biological processes and molecular functions

#GOs for biological process (BP)
GO_BP_term2name <- GO_annot_all %>% 
  dplyr::filter(ontology=="BP") %>% 
  dplyr::select(GO_ID, GO_desc) %>%
  dplyr::rename(from=GO_ID, to=GO_desc)


eggnog_to_GO_BP_annot <- eggnog_to_GO_all_annot %>% dplyr::filter(ontology == "BP") 
  
eggnog_to_GO_BP_term2gene <- eggnog_to_go_all_term2gene %>% dplyr::filter(from %in% eggnog_to_GO_BP_annot$GO_ID)

#GOs for Molecular Function (MF)
GO_MF_term2name <- GO_annot_all %>% 
  dplyr::filter(ontology=="MF") %>% 
  dplyr::select(GO_ID, GO_desc) %>%
  dplyr::rename(from=GO_ID, to=GO_desc)

eggnog_to_GO_MF_annot <- eggnog_to_GO_all_annot %>% dplyr::filter(ontology == "MF")
  
eggnog_to_GO_MF_term2gene <- eggnog_to_go_all_term2gene %>% dplyr::filter(from %in% eggnog_to_GO_MF_annot$GO_ID)


#Load KEGG pathway and signature modules metadata (e.g https://www.genome.jp/brite/ko00002)

KEGG_module_metadata <- read_tsv("../metadata/KEGG_pathway_modules_metadata",
                                 show_col_types = FALSE,
                                 col_names = c("module", "module_desc"))


#Note the technical replicates besides the 3 bact spike ins

tech_reps <- metadata %>% dplyr::filter(comments == "technical replicate analysis" & subj_region != "3 Bac (VV:LM:PS)")

mtx_to_pull_with_reps <- c(mtx_to_pull, tech_reps %>% pull(LIBID))


```

# Differential expression analysis with DESeq2, at orthologous group (OG) level

The steps taken here are:

Starting with matching pairs of MTX and MTG libraries
1) Summarize read counts for each gene feature. A gene feature could be a bacterial OG or a fungal OG. Bacterial OGs are normalized separately from fungal OGs

2) Perform pair-wise comparisons  with DESeq2: https://www.biostars.org/p/325009/#325036 & https://support.bioconductor.org/p/60032/#60055

e.g.
Scalp vs Ch
Scalp vs Vf
Scalp vs Ac
Scalp vs Tw
Ch vs Vf
Ch vs Ac
Ch vs Tw
Vf vs Ac
Vf vs Tw
Tw vs Ac


###  Step 1: Summarize read counts for each feature (IHSMGC Pangene or UniRef90 cluster)

We want to both have raw read counts and an RPK equivalent. We will not use zero smoothing/imputation. 

To calculate TPM:
Divide the read counts by the length of each gene in kilobases. This gives you reads per kilobase (RPK).
Count up all the RPK values in a sample and divide this number by 1,000,000. This is your “per million” scaling factor.
Divide the RPK values by the “per million” scaling factor. This gives you TPM.

Caveat: We will only calculate this "per million" scaling factor at the end by summing up the RPKs of all annotated features and an "inferred" RPK from UNMAPPED features (microbial reads only)

From the guidelines of https://github.com/biobakery/humann: Since other gene features in the table are quantified in RPK units, "UNMAPPED" can be interpreted as a single unknown gene of length 1 kilobase recruiting all reads that failed to map to known sequences.



```{r message=FALSE}

mtx_bact_counts <- lapply(mtx_to_pull_with_reps, function(x){
  df <- read_tsv(paste0("../data/MTX_feature_summary/",x,"_bact_OG_counts.tsv"),show_col_types = FALSE) %>% 
    dplyr::rename(feature=bacteria_OG)
  
  df$feature <- gsub(pattern="\\|Bacteria", replacement="", x=df$feature)
  
  return(df)
})

names(mtx_bact_counts) <- mtx_to_pull_with_reps

mgx_bact_counts <- lapply(mgx_to_pull, function(x){
  df <- read_tsv(paste0("../data/MGX_feature_summary/",x,"_bact_OG_counts.tsv"),show_col_types = FALSE) %>%
    dplyr::rename(feature=bacteria_OG)
  
  df$feature <- gsub(pattern="\\|Bacteria", replacement="", x=df$feature)
  
  return(df)
})

names(mgx_bact_counts) <- mgx_to_pull

#################
mtx_fungi_counts <- lapply(mtx_to_pull_with_reps, function(x){
  df <- read_tsv(paste0("../data/MTX_feature_summary/",x,"_fungi_OG_counts.tsv"),show_col_types = FALSE) %>% 
    dplyr::rename(feature=fungi_OG)
  
  df$feature <- gsub(pattern="\\|Fungi", replacement="", x=df$feature)
  
  return(df)
  
})

names(mtx_fungi_counts) <- mtx_to_pull_with_reps


mgx_fungi_counts <- lapply(mgx_to_pull, function(x){
  df <- read_tsv(paste0("../data/MGX_feature_summary/",x,"_fungi_OG_counts.tsv"),show_col_types = FALSE) %>%
    dplyr::rename(feature=fungi_OG)
  
  df$feature <- gsub(pattern="\\|Fungi", replacement="", x=df$feature)
  
  return(df)
  
})

names(mgx_fungi_counts) <- mgx_to_pull

#################
mtx_virus_counts <- lapply(mtx_to_pull_with_reps, function(x){
  read_tsv(paste0("../data/MTX_feature_summary/",x,"_virus_OG_counts.tsv"),show_col_types = FALSE) %>% 
    dplyr::rename(feature=virus_OG)
})

names(mtx_virus_counts) <- mtx_to_pull_with_reps


mgx_virus_counts <- lapply(mgx_to_pull, function(x){
  read_tsv(paste0("../data/MGX_feature_summary/",x,"_virus_OG_counts.tsv"),show_col_types = FALSE) %>%
    dplyr::rename(feature=virus_OG)
})

names(mgx_virus_counts) <- mgx_to_pull


```


### Step2: DESeq2 analysis for bacterial and fungal OGs, for the 102 paired mtx and mgx libraries

```{r}
eligible_mtx_metadata <- metadata %>% dplyr::filter(LIBID %in% c(mtx_to_pull, mgx_to_pull))

site_comparisons <- c("Sc_vs_Vf",
                      "Sc_vs_Ch",
                      "Sc_vs_Ac",
                      "Sc_vs_Tw",
                      "Ch_vs_Vf",
                      "Ch_vs_Ac",
                      "Ch_vs_Tw",
                      "Vf_vs_Ac",
                      "Tw_vs_Vf",
                      "Tw_vs_Ac")

```


Caution: If the default "ratio" method is used, many comparisons will fail and an error will be returned during estimateSizeFactors: "every gene contains at least one zero, cannot compute log geometric means". 

It needs to have additional pseudo-counts: https://www.biostars.org/p/440379/ or some form of zero handling

It is recommended to use estimateSizeFactors type = "poscounts" https://support.bioconductor.org/p/100201/ for zero inflated metagenomic data.


This is quite computationally intensive. The *.RDS files are on figshare: https://figshare.com/articles/dataset/MGX_annotations_MTX_annotations_and_pairwise_DESeq2_comparisons_at_the_orthologous_group_level/25688505

```{r eval=FALSE}

DESeq_Sc_vs_Vf <- run_DESeq_comparisons(site_to_compare = "Sc", site_reference = "Vf")
saveRDS(DESeq_Sc_vs_Vf, file="../data/DESeq2_out/DESeq_Sc_vs_Vf.RDS")

DESeq_Sc_vs_Ch <- run_DESeq_comparisons(site_to_compare = "Sc", site_reference = "Ch")
saveRDS(DESeq_Sc_vs_Ch, file="../data/DESeq2_out/DESeq_Sc_vs_Ch.RDS")

DESeq_Sc_vs_Tw <- run_DESeq_comparisons(site_to_compare = "Sc", site_reference = "Tw")
saveRDS(DESeq_Sc_vs_Tw, file="../data/DESeq2_out/DESeq_Sc_vs_Tw.RDS")

DESeq_Ch_vs_Vf <- run_DESeq_comparisons(site_to_compare = "Ch", site_reference = "Vf")
saveRDS(DESeq_Ch_vs_Vf, file="../data/DESeq2_out/DESeq_Ch_vs_Vf.RDS")

DESeq_Ch_vs_Tw <- run_DESeq_comparisons(site_to_compare = "Ch", site_reference = "Tw")
saveRDS(DESeq_Ch_vs_Tw, file="../data/DESeq2_out/DESeq_Ch_vs_Tw.RDS")

#DESeq_Vf_vs_Tw <- run_DESeq_comparisons(site_to_compare = "Vf", site_reference = "Tw")
#saveRDS(DESeq_Vf_vs_Tw, file="../data/DESeq2_out/DESeq_Vf_vs_Tw.RDS")

DESeq_Tw_vs_Vf <- run_DESeq_comparisons(site_to_compare = "Tw", site_reference = "Vf")
saveRDS(DESeq_Tw_vs_Vf, file="../data/DESeq2_out/DESeq_Tw_vs_Vf.RDS")

DESeq_Sc_vs_Ac <- run_DESeq_comparisons(site_to_compare = "Sc", site_reference = "Ac")
saveRDS(DESeq_Sc_vs_Ac, file="../data/DESeq2_out/DESeq_Sc_vs_Ac.RDS")

DESeq_Vf_vs_Ac <- run_DESeq_comparisons(site_to_compare = "Vf", site_reference = "Ac")
saveRDS(DESeq_Vf_vs_Ac, file="../data/DESeq2_out/DESeq_Vf_vs_Ac.RDS")

DESeq_Tw_vs_Ac <- run_DESeq_comparisons(site_to_compare = "Tw", site_reference = "Ac")
saveRDS(DESeq_Tw_vs_Ac, file="../data/DESeq2_out/DESeq_Tw_vs_Ac.RDS")

DESeq_Ch_vs_Ac <- run_DESeq_comparisons(site_to_compare = "Ch", site_reference = "Ac")
saveRDS(DESeq_Ch_vs_Ac, file="../data/DESeq2_out/DESeq_Ch_vs_Ac.RDS")


```

Save the log2FC results. 

```{r eval=FALSE}

write_tsv(DESeq_Sc_vs_Vf$bact_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/bact_DESeq_Sc_vs_Vf_l2fc.tsv")

write_tsv(DESeq_Sc_vs_Ch$bact_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/bact_DESeq_Sc_vs_Ch_l2fc.tsv")

write_tsv(DESeq_Sc_vs_Ac$bact_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/bact_DESeq_Sc_vs_Ac_l2fc.tsv")

write_tsv(DESeq_Sc_vs_Tw$bact_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/bact_DESeq_Sc_vs_Tw_l2fc.tsv")

write_tsv(DESeq_Ch_vs_Vf$bact_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/bact_DESeq_Ch_vs_Vf_l2fc.tsv")

write_tsv(DESeq_Ch_vs_Ac$bact_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/bact_DESeq_Ch_vs_Ac_l2fc.tsv")

write_tsv(DESeq_Ch_vs_Tw$bact_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/bact_DESeq_Ch_vs_Tw_l2fc.tsv")

write_tsv(DESeq_Vf_vs_Ac$bact_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/bact_DESeq_Vf_vs_Ac_l2fc.tsv")

#write_tsv(DESeq_Vf_vs_Tw$bact_OG_DE$DE_results_clean_shrunken_df, #file="../data/DESeq2_out/bact_DESeq_Vf_vs_Tw_l2fc.tsv")

write_tsv(DESeq_Tw_vs_Vf$bact_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/bact_DESeq_Tw_vs_Vf_l2fc.tsv")

write_tsv(DESeq_Tw_vs_Ac$bact_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/bact_DESeq_Tw_vs_Ac_l2fc.tsv")

#Fungi

write_tsv(DESeq_Sc_vs_Vf$fungal_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/fungi_DESeq_Sc_vs_Vf_l2fc.tsv")

write_tsv(DESeq_Sc_vs_Ch$fungal_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/fungi_DESeq_Sc_vs_Ch_l2fc.tsv")

write_tsv(DESeq_Sc_vs_Ac$fungal_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/fungi_DESeq_Sc_vs_Ac_l2fc.tsv")

write_tsv(DESeq_Sc_vs_Tw$fungal_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/fungi_DESeq_Sc_vs_Tw_l2fc.tsv")

write_tsv(DESeq_Ch_vs_Vf$fungal_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/fungi_DESeq_Ch_vs_Vf_l2fc.tsv")

write_tsv(DESeq_Ch_vs_Ac$fungal_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/fungi_DESeq_Ch_vs_Ac_l2fc.tsv")

write_tsv(DESeq_Ch_vs_Tw$fungal_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/fungi_DESeq_Ch_vs_Tw_l2fc.tsv")

write_tsv(DESeq_Vf_vs_Ac$fungal_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/fungi_DESeq_Vf_vs_Ac_l2fc.tsv")

#write_tsv(DESeq_Vf_vs_Tw$fungal_OG_DE$DE_results_clean_shrunken_df, #file="../data/DESeq2_out/fungi_DESeq_Vf_vs_Tw_l2fc.tsv")

write_tsv(DESeq_Tw_vs_Vf$fungal_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/fungi_DESeq_Tw_vs_Vf_l2fc.tsv")

write_tsv(DESeq_Tw_vs_Ac$fungal_OG_DE$DE_results_clean_shrunken_df, file="../data/DESeq2_out/fungi_DESeq_Tw_vs_Ac_l2fc.tsv")
```

Gene set enrichment analysis based on log2 fold change results from these pairwise comparisons

```{r eval=FALSE}

#This takes quite long, so save the results and load them separately
site_DE_res_enrichments_GO_MF_GSEA <- lapply(site_comparisons, function(x){
  
  list_out <- tally_features_and_enrichments(x, 
                                             term2gene_arg = eggnog_to_GO_MF_term2gene,
                                             term2name_arg = GO_MF_term2name,
                                            analysis_mode = "GSEA")
  
  return(list_out)
  
})

names(site_DE_res_enrichments_GO_MF_GSEA) <- site_comparisons

saveRDS(site_DE_res_enrichments_GO_MF_GSEA, file="../data/DESeq2_out/site_DE_res_enrichments_GO_MF_GSEA.RDS")

##BP (biological process)
site_DE_res_enrichments_GO_BP_GSEA <- lapply(site_comparisons, function(x){
  
  list_out <- tally_features_and_enrichments(x, 
                                             term2gene_arg = eggnog_to_GO_BP_term2gene,
                                             term2name_arg = GO_BP_term2name,
                                            analysis_mode = "GSEA")
  
  return(list_out)
  
})

names(site_DE_res_enrichments_GO_BP_GSEA) <- site_comparisons

saveRDS(site_DE_res_enrichments_GO_BP_GSEA, file="../data/DESeq2_out/site_DE_res_enrichments_GO_BP_GSEA.RDS")

##What if we did GSEA with KEGG pathways
site_DE_res_enrichments_KEGG_GSEA <- lapply(site_comparisons, function(x){
  
  list_out <- tally_features_and_enrichments(x, 
                                             term2gene_arg = bacteria_and_fungi_eggnog_to_KEGG_term2gene,
                                             term2name_arg = KEGG_term2name,
                                            analysis_mode = "GSEA")
  
  return(list_out)
  
})


```

Load the GSEA results from the previous code chunk.

```{r}

site_DE_res_enrichments_GO_MF_GSEA <- readRDS(file="../data/DESeq2_out/site_DE_res_enrichments_GO_MF_GSEA.RDS")

site_DE_res_enrichments_GO_BP_GSEA <- readRDS(file="../data/DESeq2_out/site_DE_res_enrichments_GO_BP_GSEA.RDS")


```

Load GSEA results from aldex2 for the same samples.

Here, the abundances of each taxon were subjected to centered log ratio (clr) transform, averaged over 1000 MC instances

Clr values were used as inputs into Maaslin2 as a mixed effects model, to find RNAs enriched in one skin site relative to another, while accounting for variations in DNA abundances.

Log fold changes were used an inputs for GSEA

```{r}

site_clr_enrichments_GO_MF_GSEA <- readRDS(file="../data/aldex2_glmm_out/site_clr_enrichments_GO_MF_GSEA.RDS")

site_clr_enrichments_GO_BP_GSEA <- readRDS(file="../data/aldex2_glmm_out/site_clr_enrichments_GO_BP_GSEA.RDS")

```

Load GSEA results from ANCOMBC2 for the same samples.

ANCOMBC formulation explicitly tests the hypothesis regarding differential absolute (not relative!) abundance of individual taxon while estimating sample-specific sampling fractions and correcting the bias appropriately.

```{r}
site_ANCOMBC_res_enrichments_GO_MF_GSEA <-readRDS(file="../data/ancombc2_out/site_ANCOMBC_res_enrichments_GO_MF_GSEA.RDS")

site_ANCOMBC_res_enrichments_GO_BP_GSEA <- readRDS(file="../data/ancombc2_out/site_ANCOMBC_res_enrichments_GO_BP_GSEA.RDS")

```

Plot NES results only if independently picked up by 2 out of the 3 methods

Change NES bar plot to NES dot plots. Each dot being a different color based on the method.

```{r}


summarise_GSEA_helper <- function(DESeq_GSEA_in, 
                                  aldex_GSEA_in,
                                  ancombc_GSEA_in){
  
  DESeq_GSEA_in$signif_GSEA_DESeq2 <- TRUE
  
  aldex_GSEA_in$signif_GSEA_clr <- TRUE
  
  ancombc_GSEA_in$signif_GSEA_ancombc <- TRUE  
  
  ##Number of times a feature is significantly enriched by GSEA amongst the three methods
  
  feature_GSEA_tally <- rbind(DESeq_GSEA_in[,c("ID", "Description")], 
                              aldex_GSEA_in[,c("ID", "Description")], 
                              ancombc_GSEA_in[,c("ID", "Description")])
  
  feature_GSEA_tally_count <- plyr::count(feature_GSEA_tally, c("ID", "Description"))
  
  ##Add information about which methods were statistically significant
  #https://stackoverflow.com/questions/8091303/simultaneously-merge-multiple-data-frames-in-a-list
  
  feature_GSEA_tally_count_fmt <- list(feature_GSEA_tally_count, 
                                DESeq_GSEA_in[,c("ID", "signif_GSEA_DESeq2", "DESeq_GSEA_NES", "DESeq_GSEA_padj")],
                                aldex_GSEA_in[,c("ID","signif_GSEA_clr", "aldex_GSEA_NES", "aldex_GSEA_padj")],
                                ancombc_GSEA_in[,c("ID","signif_GSEA_ancombc", "ancombc_GSEA_NES", "ancombc_GSEA_padj")]) %>% 
  purrr::reduce(left_join, by = "ID")
  
  feature_GSEA_tally_count_fmt <- feature_GSEA_tally_count_fmt %>%
  mutate(across(starts_with("signif"), ~ replace_na(.x, FALSE)))
  
  return(feature_GSEA_tally_count_fmt)
  
}


#add dummy row to an empty dataframe for merging purposes
add_dummy <- function(df){
  
  if(nrow(df)==0){
    
    df [ nrow(df) + 1 , ] <- NA
    df$ID <- "placeholder"
  }
  
  return(df)
  
}


#e.g. site_DE_res_enrichments_GO_BP_GSEA
#comparison can be "Ch_vs_Vf" etc
compare_GSEA_across <- function(DESeq_GSEA_res, 
                                aldex_GSEA_res, 
                                ancombc_GSEA_res,
                                comparison){
  
  ##############
  #For bacteria#
  ##############
  bact_DESeq_GSEA <- DESeq_GSEA_res[[comparison]]$bact_DESeq_GSEA_results@result %>% 
    dplyr::rename("DESeq_GSEA_NES"="NES", "DESeq_GSEA_padj"="p.adjust") %>% add_dummy(.)
  
  bact_aldex_GSEA <- aldex_GSEA_res[[comparison]]$bact_GSEA_results@result %>% 
    dplyr::rename("aldex_GSEA_NES"="NES", "aldex_GSEA_padj"="p.adjust") %>% add_dummy(.)
  
  bact_ancombc_GSEA <- ancombc_GSEA_res[[comparison]]$bact_GSEA_results@result %>% 
    dplyr::rename("ancombc_GSEA_NES"="NES", "ancombc_GSEA_padj"="p.adjust") %>% add_dummy(.)
  
  #tally and summarize GSEA results across the three methods
  bact_GSEA_tally <- summarise_GSEA_helper(DESeq_GSEA_in=bact_DESeq_GSEA,
                                           aldex_GSEA_in=bact_aldex_GSEA,
                                           ancombc_GSEA_in=bact_ancombc_GSEA) %>% 
    dplyr:::filter(ID != "placeholder")
  
  
  #for plotting those pathways with at least two hits from distinct methods
  bact_GSEA_tally_for_plot_padj <- bact_GSEA_tally %>% dplyr::filter(freq >= 2) %>% 
    dplyr::select(c("ID", "Description",
                  "DESeq_GSEA_padj", 
                  "aldex_GSEA_padj", 
                  "ancombc_GSEA_padj")) %>% pivot_longer(!c("ID","Description"), 
                                                         names_to = "method", 
                                                         values_to = "GSEA_padj")
  
  bact_GSEA_tally_for_plot_padj$method <- gsub(pattern="_padj", replacement="", x=bact_GSEA_tally_for_plot_padj$method)
  
  bact_GSEA_tally_for_plot_NES <- bact_GSEA_tally %>% dplyr::filter(freq >= 2) %>% 
    dplyr::select(c("ID",
                  "DESeq_GSEA_NES", 
                  "aldex_GSEA_NES", 
                  "ancombc_GSEA_NES")) %>% pivot_longer(!ID, 
                                                         names_to = "method", 
                                                         values_to = "GSEA_NES")
  
  bact_GSEA_tally_for_plot_NES$method <- gsub(pattern="_NES", replacement="", x=bact_GSEA_tally_for_plot_NES$method)
  
  
  bact_GSEA_tally_for_plot <- merge(bact_GSEA_tally_for_plot_padj, bact_GSEA_tally_for_plot_NES, 
                                    by = c("ID","method"))
  
  ###########
  #For fungi#
  ###########
  
  fungi_DESeq_GSEA <- DESeq_GSEA_res[[comparison]]$fungi_DESeq_GSEA_results@result %>% 
    dplyr::rename("DESeq_GSEA_NES"="NES", "DESeq_GSEA_padj"="p.adjust") %>% add_dummy(.)
  
  fungi_aldex_GSEA <- aldex_GSEA_res[[comparison]]$fungi_GSEA_results@result %>% 
    dplyr::rename("aldex_GSEA_NES"="NES", "aldex_GSEA_padj"="p.adjust") %>% add_dummy(.)
  
  fungi_ancombc_GSEA <- ancombc_GSEA_res[[comparison]]$fungi_GSEA_results@result %>% 
    dplyr::rename("ancombc_GSEA_NES"="NES", "ancombc_GSEA_padj"="p.adjust") %>% add_dummy(.)
  
  #tally and summarize GSEA results across the three methods
  fungi_GSEA_tally <- summarise_GSEA_helper(DESeq_GSEA_in=fungi_DESeq_GSEA,
                                           aldex_GSEA_in=fungi_aldex_GSEA,
                                           ancombc_GSEA_in=fungi_ancombc_GSEA) %>% 
    dplyr:::filter(ID != "placeholder")
  
  
  fungi_GSEA_tally_for_plot_padj <- fungi_GSEA_tally %>% dplyr::filter(freq >= 2) %>% 
    dplyr::select(c("ID", "Description",
                  "DESeq_GSEA_padj", 
                  "aldex_GSEA_padj", 
                  "ancombc_GSEA_padj")) %>% pivot_longer(!c("ID","Description"), 
                                                         names_to = "method", 
                                                         values_to = "GSEA_padj")
  
  fungi_GSEA_tally_for_plot_padj$method <- gsub(pattern="_padj", replacement="", x=fungi_GSEA_tally_for_plot_padj$method)
  
  fungi_GSEA_tally_for_plot_NES <- fungi_GSEA_tally %>% dplyr::filter(freq >= 2) %>% 
    dplyr::select(c("ID",
                  "DESeq_GSEA_NES", 
                  "aldex_GSEA_NES", 
                  "ancombc_GSEA_NES")) %>% pivot_longer(!ID, 
                                                         names_to = "method", 
                                                         values_to = "GSEA_NES")
  
  fungi_GSEA_tally_for_plot_NES$method <- gsub(pattern="_NES", replacement="", x=fungi_GSEA_tally_for_plot_NES$method)
  
  
  fungi_GSEA_tally_for_plot <- merge(fungi_GSEA_tally_for_plot_padj, fungi_GSEA_tally_for_plot_NES, 
                                    by = c("ID","method"))
  
  
  output <- tibble::lst(bact_GSEA_tally, bact_GSEA_tally_for_plot,
                        fungi_GSEA_tally, fungi_GSEA_tally_for_plot)
  
  return(output)
  
}


```

Tally the scores across the three methods,

For Ch versus Vf

```{r}
#for BP
Ch_vs_Vf_GSEA_GO_BP_multi <- compare_GSEA_across(DESeq_GSEA_res = site_DE_res_enrichments_GO_BP_GSEA,
                              aldex_GSEA_res = site_clr_enrichments_GO_BP_GSEA,
                              ancombc_GSEA_res = site_ANCOMBC_res_enrichments_GO_BP_GSEA,
                              comparison="Ch_vs_Vf")


#For MF
Ch_vs_Vf_GSEA_GO_MF_multi <- compare_GSEA_across(DESeq_GSEA_res = site_DE_res_enrichments_GO_MF_GSEA,
                              aldex_GSEA_res = site_clr_enrichments_GO_MF_GSEA,
                              ancombc_GSEA_res = site_ANCOMBC_res_enrichments_GO_MF_GSEA,
                              comparison="Ch_vs_Vf")



```

For Tw versus Vf

```{r}

#for BP
Tw_vs_Vf_GSEA_GO_BP_multi <- compare_GSEA_across(DESeq_GSEA_res = site_DE_res_enrichments_GO_BP_GSEA,
                              aldex_GSEA_res = site_clr_enrichments_GO_BP_GSEA,
                              ancombc_GSEA_res = site_ANCOMBC_res_enrichments_GO_BP_GSEA,
                              comparison="Tw_vs_Vf")

#For MF
Tw_vs_Vf_GSEA_GO_MF_multi <- compare_GSEA_across(DESeq_GSEA_res = site_DE_res_enrichments_GO_MF_GSEA,
                              aldex_GSEA_res = site_clr_enrichments_GO_MF_GSEA,
                              ancombc_GSEA_res = site_ANCOMBC_res_enrichments_GO_MF_GSEA,
                              comparison="Tw_vs_Vf")



```

## Normalized enrichment score plots

(Panel A) Barplots of normalized enrichment scores (NES) computed from gene set enrichment analyses of bacterial or fungal orthologous groups that are
differentially enriched between cheek (n=22) and volar forearm (n=18) metatranscriptomes. 

All pathways presented here were statistically significant across at least two methods (DESeq2, aldex clr - glmm and ANCOM-BC).

```{r}

#################
###GO Slimming###
#################


##MF##
Ch_vs_Vf_bact_GO_MF_GSEA_slim <- associate_GO_to_slim(Ch_vs_Vf_GSEA_GO_MF_multi$bact_GSEA_tally_for_plot$ID %>% unique(.), mode="MF")

Ch_vs_Vf_bact_GO_MF_GSEA_fmt <- prepare_NES_plot_input(df=Ch_vs_Vf_GSEA_GO_MF_multi$bact_GSEA_tally_for_plot,
                                                               slimdf=Ch_vs_Vf_bact_GO_MF_GSEA_slim,
                                                               chosen_parent_ids = "nucleic acid binding")

Ch_vs_Vf_bact_GO_MF_GSEA_fmt$organism <- "Bacteria"
Ch_vs_Vf_bact_GO_MF_GSEA_fmt$GO_level <- "Molecular function"

#

Ch_vs_Vf_fungi_GO_MF_GSEA_slim <- associate_GO_to_slim(Ch_vs_Vf_GSEA_GO_MF_multi$fungi_GSEA_tally_for_plot$ID %>% unique(.), mode="MF")

Ch_vs_Vf_fungi_GO_MF_GSEA_fmt <- prepare_NES_plot_input(df=Ch_vs_Vf_GSEA_GO_MF_multi$fungi_GSEA_tally_for_plot,
                                                               slimdf=Ch_vs_Vf_fungi_GO_MF_GSEA_slim,
                                                               chosen_parent_ids = c("nucleic acid binding",
                                                                                     "antioxidant activity"))

Ch_vs_Vf_fungi_GO_MF_GSEA_fmt$organism <- "Fungi"
Ch_vs_Vf_fungi_GO_MF_GSEA_fmt$GO_level <- "Molecular function"





##BP##

Ch_vs_Vf_bact_GO_BP_GSEA_slim <- associate_GO_to_slim(Ch_vs_Vf_GSEA_GO_BP_multi$bact_GSEA_tally_for_plot$ID %>% unique(.), mode="BP")

Ch_vs_Vf_bact_GO_BP_GSEA_fmt <- prepare_NES_plot_input(df=Ch_vs_Vf_GSEA_GO_BP_multi$bact_GSEA_tally_for_plot,
                                                               slimdf=Ch_vs_Vf_bact_GO_BP_GSEA_slim,
                                                               chosen_parent_ids = c(
                                                       "transport", "translation", "carbohydrate metabolic process",
                                                       "protein-containing complex assembly",
                                                       "generation of precursor metabolites and energy"))

Ch_vs_Vf_bact_GO_BP_GSEA_fmt$organism <- "Bacteria"
Ch_vs_Vf_bact_GO_BP_GSEA_fmt$GO_level <- "Biological process"

#

Ch_vs_Vf_fungi_GO_BP_GSEA_slim <- associate_GO_to_slim(Ch_vs_Vf_GSEA_GO_BP_multi$fungi_GSEA_tally_for_plot$ID %>% unique(.), mode="BP")

Ch_vs_Vf_fungi_GO_BP_GSEA_fmt <- prepare_NES_plot_input(df=Ch_vs_Vf_GSEA_GO_BP_multi$fungi_GSEA_tally_for_plot,
                                                        slimdf=Ch_vs_Vf_fungi_GO_BP_GSEA_slim,
                                                        chosen_parent_ids = c("DNA conformation change"))

Ch_vs_Vf_fungi_GO_BP_GSEA_fmt$organism <- "Fungi"
Ch_vs_Vf_fungi_GO_BP_GSEA_fmt$GO_level <- "Biological process"

########################################################
#Combine the bacteria and fungi dataframes for plotting#
########################################################

#All the enriched functions for bact MF are to do with ribosome assembly
Ch_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df <- Ch_vs_Vf_bact_GO_MF_GSEA_fmt %>% 
  dplyr::filter(Description %in% c("structural constituent of ribosome",
                                   "rRNA binding",
                                   "large ribosomal subunit rRNA binding"))

#
#only hits to "antioxidant activity"
Ch_vs_Vf_fungi_GO_MF_GSEA_NES_condensed_df <- Ch_vs_Vf_fungi_GO_MF_GSEA_fmt 


#
Ch_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df <- Ch_vs_Vf_bact_GO_BP_GSEA_fmt %>% 
                                                                            dplyr::filter(Description %in% 
                                                                                            c("glucose catabolic process",
                                                                                              "canonical glycolysis",
                                                                                              "proton motive force-driven ATP synthesis",
                                                                                              "NADH regeneration",
                                                                                              "cotranslational protein targeting to membrane",
                                                                                              "ribosomal large subunit assembly",
                                                                                              "ribosomal small subunit assembly"))


Ch_vs_Vf_fungi_GO_BP_GSEA_NES_condensed_df <- Ch_vs_Vf_fungi_GO_BP_GSEA_fmt %>%
                                                                              dplyr::filter(Description %in% 
                                                                                              c("DNA geometric change",
                                                                                           "DNA duplex unwinding",
                                                                                            "DNA conformation change"))

##Add protein modification process (GO:0036211) and protein lipidation (GO:0006497) to the plot (as extras). The former is a parent term of the latter
#This is for biological process for fungi

#protein modification process (GO:0036211) and protein lipidation (GO:0006497)
Ch_vs_Vf_GSEA_GO_BP_fungi_GSEA_extras <- extract_NES_stats(df =Ch_vs_Vf_GSEA_GO_BP_multi$fungi_GSEA_tally, 
                                                           GO_description = c("protein modification process", "protein lipidation")) %>%
  dplyr::filter(ID %in% c("GO:0036211", "GO:0006497"))

Ch_vs_Vf_GSEA_GO_BP_fungi_GSEA_extras$GO_slim_parent_term <- NA
Ch_vs_Vf_GSEA_GO_BP_fungi_GSEA_extras$organism <- "Fungi"
Ch_vs_Vf_GSEA_GO_BP_fungi_GSEA_extras$GO_level <- "Biological process"

##
Ch_vs_Vf_all_GO_GSEA_NES_condensed_df <- do.call("rbind", lst(Ch_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df,
                                                              Ch_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df,
                                                              Ch_vs_Vf_fungi_GO_BP_GSEA_NES_condensed_df,
                                                              Ch_vs_Vf_fungi_GO_MF_GSEA_NES_condensed_df,
                                                              Ch_vs_Vf_GSEA_GO_BP_fungi_GSEA_extras))


Ch_vs_Vf_all_GO_GSEA_NES_condensed_df$direction <- ifelse(Ch_vs_Vf_all_GO_GSEA_NES_condensed_df$GSEA_NES >0,
                                                          "enriched_in_Ch", "enriched_in_Vf")



Ch_vs_Vf_all_GO_GSEA_NES_condensed_plot <- ggplot(Ch_vs_Vf_all_GO_GSEA_NES_condensed_df, 
                                                  aes(y=Description, x=GSEA_NES, color = method)) +
  geom_point(position="dodge", size =5, alpha = 0.5) +  
  geom_vline(xintercept=0, linetype="longdash") +
  scale_fill_manual(values = c23) +
  theme(axis.text.y = element_text(vjust = 0.5, hjust=1)) +
  labs(x="Normalized Enrichment Score") + 
   facet_grid(rows=vars(organism),cols=vars(GO_level),
             space = "free_y", scales="free_y",
             labeller = label_wrap_gen(width = 1.5, multi_line = TRUE)) + 
  coord_cartesian(xlim = c(-4.6, 2.6)) +
  theme_classic() +
  theme(panel.spacing = unit(.05, "lines"),
        panel.border = element_rect(color = "black", fill = NA, linewidth = 0.2), 
       strip.background = element_rect(color = "black", linewidth = 1)) + ggtitle("Cheeks vs. Volar forearms")


Ch_vs_Vf_all_GO_GSEA_NES_condensed_plot


```


(Panel B) Same as (A)
but computed between toe web (n=18) and volar forearm (n=18) metatranscriptomes. Biologically noteworthy pathways are bolded with an asterisk for clarity.

All pathways presented here were statistically significant across at least two methods (DESeq2, aldex clr - glmm and ANCOM-BC).

This comparison only makes sense for bacterial OGs/gene clusters because fungal RNAs are very scarce on toe webs.

```{r}

#################
###GO Slimming###
#################



##MF##
Tw_vs_Vf_bact_GO_MF_GSEA_slim <- associate_GO_to_slim(Tw_vs_Vf_GSEA_GO_MF_multi$bact_GSEA_tally_for_plot$ID %>% unique(.), mode="MF")

Tw_vs_Vf_bact_GO_MF_GSEA_fmt <- prepare_NES_plot_input(df=Tw_vs_Vf_GSEA_GO_MF_multi$bact_GSEA_tally_for_plot,
                                                               slimdf=Tw_vs_Vf_bact_GO_MF_GSEA_slim,
                                                               chosen_parent_ids = c("catalytic activity",
                                                                                     "transferase activity",
                                                                                     "hydrolase activity",
                                                                                     "lyase activity",
                                                                                     "ligase activity",
                                                                                     "oxidoreductase activity",
                                                                                     "antioxidant activity"))

Tw_vs_Vf_bact_GO_MF_GSEA_fmt$organism <- "Bacteria"
Tw_vs_Vf_bact_GO_MF_GSEA_fmt$GO_level <- "Molecular function"

#

Tw_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df <- Tw_vs_Vf_bact_GO_MF_GSEA_fmt %>%
                                                                              dplyr::filter(Description %in% 
                                                                                              c("ribonucleotide binding",
                                                                                            "DNA helicase activity",
                                                                                            "exonuclease activity",
                                                                                            "RNA methyltransferase activity","ligase activity, forming carbon-nitrogen bonds",
"hydrolase activity, acting on ester bonds",
"kinase activity",
"antioxidant activity",
"peroxidase activity",
"lyase activity"))



##BP##

Tw_vs_Vf_bact_GO_BP_GSEA_slim <- associate_GO_to_slim(Tw_vs_Vf_GSEA_GO_BP_multi$bact_GSEA_tally_for_plot$ID %>% unique(.), mode="BP")

Tw_vs_Vf_bact_GO_BP_GSEA_fmt <- prepare_NES_plot_input(df=Tw_vs_Vf_GSEA_GO_BP_multi$bact_GSEA_tally_for_plot,
                                                               slimdf=Tw_vs_Vf_bact_GO_BP_GSEA_slim,
                                                               chosen_parent_ids = c("nitrogen compound metabolic process",
                                                                                     "RNA metabolic process",
                                                                                     "DNA metabolic process",
                                                                                     "biosynthetic process",
                                                                                     "transport",
                                                                                     "small molecule metabolic process"))

Tw_vs_Vf_bact_GO_BP_GSEA_fmt$organism <- "Bacteria"
Tw_vs_Vf_bact_GO_BP_GSEA_fmt$GO_level <- "Biological process"


Tw_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df <- Tw_vs_Vf_bact_GO_BP_GSEA_fmt %>% 
                                                                            dplyr::filter(Description %in% 
                                                                                            c("tricarboxylic acid cycle",
"cell cycle DNA replication",
"amino acid metabolic process",
"thiamine biosynthetic process",
"protoporphyrinogen IX metabolic process",
"RNA methylation",
"sodium ion transmembrane transport",
"amino acid transport",
"carboxylic acid transport",
"transposition"))


Tw_vs_Vf_all_GO_GSEA_NES_condensed_df <- do.call("rbind", lst(Tw_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df,
                                                              Tw_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df)) %>%
  dplyr::select(-GO_slim_parent_term) %>% unique(.)


Tw_vs_Vf_all_GO_GSEA_NES_condensed_df$direction <- ifelse(Ch_vs_Vf_all_GO_GSEA_NES_condensed_df$GSEA_NES >0,
                                                          "enriched_in_Tw", "enriched_in_Vf")


Tw_vs_Vf_all_GO_GSEA_NES_condensed_plot <- ggplot(Tw_vs_Vf_all_GO_GSEA_NES_condensed_df, 
                                                  aes(y=Description, x=GSEA_NES, color = method)) +
  geom_point(position="dodge", size =5, alpha = 0.5) +  
  geom_vline(xintercept=0, linetype="longdash") +
  scale_fill_manual(values = c23) +
  theme(axis.text.y = element_text(vjust = 0.5, hjust=1)) +
  labs(x="Normalized Enrichment Score") + 
   facet_grid(rows=vars(organism),cols=vars(GO_level),
             space = "free_y", scales="free_y",
             labeller = label_wrap_gen(width = 1.5, multi_line = TRUE)) + 
  coord_cartesian(xlim = c(-2.6, 2.6)) +
  theme_classic() +
  theme(panel.spacing = unit(.05, "lines"),
        panel.border = element_rect(color = "black", fill = NA, linewidth = 0.2), 
       strip.background = element_rect(color = "black", linewidth = 1)) + ggtitle("Toe webs vs. Volar forearms")


Tw_vs_Vf_all_GO_GSEA_NES_condensed_plot


```





Normalized Enrichment Score (NES) plots for Toe webs vs Volar forearms.

This comparison only makes sense for bacterial OGs/gene clusters because fungal RNAs are very scarce on toe webs.


```{r}
Tw_vs_Vf_bact_GO_MF_GSEA <- site_DE_res_enrichments_GO_MF_GSEA$Tw_vs_Vf$bact_DESeq_GSEA_results@result

Tw_vs_Vf_bact_GO_MF_GSEA_slim <- associate_GO_to_slim(Tw_vs_Vf_bact_GO_MF_GSEA$ID, mode="MF")

Tw_vs_Vf_bact_GO_MF_GSEA_fmt <- prepare_NES_plot_input(df=Tw_vs_Vf_bact_GO_MF_GSEA,
                                                               slimdf=Tw_vs_Vf_bact_GO_MF_GSEA_slim,
                                                               chosen_parent_ids = c("catalytic activity",
                                                                                     "transferase activity",
                                                                                     "hydrolase activity",
                                                                                     "oxidoreductase activity",
                                                                                     "antioxidant activity"))


Tw_vs_Vf_bact_GO_BP_GSEA <- site_DE_res_enrichments_GO_BP_GSEA$Tw_vs_Vf$bact_DESeq_GSEA_results@result

Tw_vs_Vf_bact_GO_BP_GSEA_slim <- associate_GO_to_slim(Tw_vs_Vf_bact_GO_BP_GSEA$ID, mode="BP")

Tw_vs_Vf_bact_GO_BP_GSEA_fmt <- prepare_NES_plot_input(df=Tw_vs_Vf_bact_GO_BP_GSEA,
                                                               slimdf=Tw_vs_Vf_bact_GO_BP_GSEA_slim,
                                                               chosen_parent_ids = c("nitrogen compound metabolic process",
                                                                                     "RNA metabolic process",
                                                                                     "DNA metabolic process",
                                                                                     "transport",
                                                                                     "small molecule metabolic process",
                                                                                     "generation of precursor metabolites and energy","response to stress","response to abiotic stimulus","response to chemical"))


Tw_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df <- Tw_vs_Vf_bact_GO_BP_GSEA_fmt %>%
                                                                              dplyr::filter(Description %in% 
                                                                                              c("tricarboxylic acid cycle",
"cell cycle DNA replication",
"amino acid metabolic process",
"thiamine biosynthetic process",
"protoporphyrinogen IX metabolic process",
"RNA methylation",
"sodium ion transmembrane transport",
"amino acid transport",
"carboxylic acid transport",
"transposition")) %>% dplyr::select(Description, NES, GO_slim_parent_term)

Tw_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df$organism <- "Bacteria"
Tw_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df$GO_level <- "Biological process"


Tw_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df <- Tw_vs_Vf_bact_GO_MF_GSEA_fmt %>%
                                                                              dplyr::filter(Description %in% 
                                                                                              c("ribonucleotide binding",
                                                                                            "DNA helicase activity",
                                                                                            "exonuclease activity",
                                                                                            "RNA methyltransferase activity","ligase activity, forming carbon-nitrogen bonds",
"hydrolase activity, acting on ester bonds",
"kinase activity",
"antioxidant activity",
"peroxidase activity",
"fibronectin binding")) %>% dplyr::select(Description, NES, GO_slim_parent_term)

Tw_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df$organism <- "Bacteria"
Tw_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df$GO_level <- "Molecular function"


Tw_vs_Vf_all_GO_GSEA_NES_condensed_df <- do.call("rbind", lst(Tw_vs_Vf_bact_GO_BP_GSEA_NES_condensed_df,
                                                              Tw_vs_Vf_bact_GO_MF_GSEA_NES_condensed_df,
                                                           ))


Tw_vs_Vf_all_GO_GSEA_NES_condensed_df$direction <- ifelse(Tw_vs_Vf_all_GO_GSEA_NES_condensed_df$NES >0,
                                                          "enriched_in_Tw", "enriched_in_Vf")




Tw_vs_Vf_all_GO_GSEA_NES_condensed_plot <- ggplot(Tw_vs_Vf_all_GO_GSEA_NES_condensed_df, aes(x=Description, y=NES, fill = direction)) +
  geom_col(colour="black",position="dodge") +  
  geom_hline(yintercept=0, linetype="longdash") +
  scale_fill_manual(values = c23) +
  theme(axis.text.x = element_text(vjust = 0.5, hjust=1)) +
  coord_flip() +
  labs(y="Normalized Enrichment Score") + 
   facet_grid(rows=vars(organism),cols=vars(GO_level),
             scales = "free", space = "free",
             labeller = label_wrap_gen(width = 1.5, multi_line = TRUE)) + 
  theme_classic() +
  theme(panel.spacing = unit(.05, "lines"),
        panel.border = element_rect(color = "black", fill = NA, size = 0.2), 
       strip.background = element_rect(color = "black", size = 1)) + ggtitle("Toe webs vs. Volar forearms")


Tw_vs_Vf_all_GO_GSEA_NES_condensed_plot



```

Assemble the panels into the final figure

```{r}

Supp_between_site_GSEA_NES_plots <- ggarrange(Ch_vs_Vf_all_GO_GSEA_NES_condensed_plot,
                                              Tw_vs_Vf_all_GO_GSEA_NES_condensed_plot,
                                              ncol =2,
                                              labels = c("A", "B"))


ggsave(filename="../plots/Supp_between_site_GSEA_NES_plots.pdf",
       plot=Supp_between_site_GSEA_NES_plots,
       height = 8, width = 18)

Supp_between_site_GSEA_NES_plots

```

# Volcano plots

(Panel A) Volcano plot of expressed bacterial and fungal orthologous groups (OGs) in cheeks (n=22) relative to volar forearms (n=18). 

(Panel B) Volcano plot of expressed bacterial orthologous groups (OGs) in toe webs (n=18) relative to volar forearms (n=18).


## Ch vs Vf

```{r}

eggnog_prot_mod_process <- eggnog_to_GO_all_annot %>% dplyr::filter(GO_desc=="protein modification process")

eggnog_glu_catab <- eggnog_to_GO_all_annot %>% dplyr::filter(GO_desc=="glucose catabolic process")

#
bact_DESeq_Ch_vs_Vf_l2fc <- read_tsv("../data/DESeq2_out/bact_DESeq_Ch_vs_Vf_l2fc.tsv", show_col_types = FALSE)
fungi_DESeq_Ch_vs_Vf_l2fc <- read_tsv("../data/DESeq2_out/fungi_DESeq_Ch_vs_Vf_l2fc.tsv", show_col_types = FALSE)


#Prepare df for volcano plot
Ch_vs_Vf_for_volcano <- site_DE_res_enrichments_GO_BP_GSEA$Ch_vs_Vf$combined_DE_res


# this can be achieved with nested ifelse statements
Ch_vs_Vf_volcano_keyvals_color <- ifelse(
  Ch_vs_Vf_for_volcano$feature %in% intersect(eggnog_prot_mod_process$OG, eggnog_glu_catab$OG), 'darkgreen',
      ifelse(Ch_vs_Vf_for_volcano$feature %in% eggnog_glu_catab$OG, 'gold',
             ifelse(Ch_vs_Vf_for_volcano$feature %in% eggnog_prot_mod_process$OG, 'royalblue',
        'black')))
Ch_vs_Vf_volcano_keyvals_color[is.na(Ch_vs_Vf_volcano_keyvals_color)] <- 'black'
names(Ch_vs_Vf_volcano_keyvals_color)[Ch_vs_Vf_volcano_keyvals_color == 'gold'] <- 'glucose catabolic process'
names(Ch_vs_Vf_volcano_keyvals_color)[Ch_vs_Vf_volcano_keyvals_color == 'royalblue'] <- 'Protein modification process'
names(Ch_vs_Vf_volcano_keyvals_color)[Ch_vs_Vf_volcano_keyvals_color == 'darkgreen'] <- 'Both'

Ch_vs_Vf_volcano_keyvals_shape <- ifelse(
  Ch_vs_Vf_for_volcano$feature %in% (Ch_vs_Vf_for_volcano %>% dplyr::filter(feature_from=="fungi") %>% pull(feature)), 17,19)

names(Ch_vs_Vf_volcano_keyvals_shape)[Ch_vs_Vf_volcano_keyvals_shape == 17] <- 'Fungi OGs'
names(Ch_vs_Vf_volcano_keyvals_shape)[Ch_vs_Vf_volcano_keyvals_shape == 19] <- 'Bacteria OGs'


#For labelling purposes
bact_DESeq_Ch_vs_Vf_GO_glu_catab <- bact_DESeq_Ch_vs_Vf_l2fc %>% dplyr::filter(feature %in% eggnog_glu_catab$OG)

fungi_DESeq_Ch_vs_Vf_GO_glu_catab <- fungi_DESeq_Ch_vs_Vf_l2fc %>% dplyr::filter(feature %in% eggnog_glu_catab$OG)

bact_DESeq_Ch_vs_Vf_GO_prot_mod_process <- bact_DESeq_Ch_vs_Vf_l2fc %>% dplyr::filter(feature %in% eggnog_prot_mod_process$OG)

fungi_DESeq_Ch_vs_Vf_GO_prot_mod_process<- fungi_DESeq_Ch_vs_Vf_l2fc %>% dplyr::filter(feature %in% eggnog_prot_mod_process$OG)

##

#Warning: the vector for the lab= argument must match the order of the rownames in the results
#best to use the same dfs as inputs
OG_Ch_vs_Vf_DE_volcano <- EnhancedVolcano(Ch_vs_Vf_for_volcano,
                lab = Ch_vs_Vf_for_volcano$feature, 
                title = "Bacteria and fungal OGs, Ch vs Vf",
                x='log2FoldChange',
                y='padj',
                #boxedLabels = TRUE,
                drawConnectors = TRUE,
                selectLab = c("COG3588", "COG0148", "COG1063",
                              "3NWH3", "3NZ4F", "3NTZT", "3NU7J", "3NUQC"),
                labSize = 3.0,
                labFace = 'bold',
                pCutoff = 0.05,
                FCcutoff=0,
                legendPosition = 'top',
              colCustom = Ch_vs_Vf_volcano_keyvals_color,
              shapeCustom = Ch_vs_Vf_volcano_keyvals_shape,
               )


ggsave(filename="../plots/OG_Ch_vs_Vf_DE_volcano.pdf",
       plot=OG_Ch_vs_Vf_DE_volcano,
       height = 8, width = 10)


OG_Ch_vs_Vf_DE_volcano

```

## Tw vs Vf

COG0407 is hemE https://www.ncbi.nlm.nih.gov/research/cog/cog/COG0407/
https://www.ncbi.nlm.nih.gov/research/cog/cog/COG1232/ is hemY
https://www.ncbi.nlm.nih.gov/research/cog/cog/COG0113/ is hemB
https://www.ncbi.nlm.nih.gov/research/cog/cog/COG1587/ is hemD
COG0008 is a glutamyl tRNA transferase
COG0001 is hemL

https://microbialcellfactories.biomedcentral.com/articles/10.1186/s12934-023-02077-3 (heme biosynthesis and pathway diagram)


```{r}

eggnog_AA_metabolic <- eggnog_to_GO_all_annot %>% dplyr::filter(GO_ID=="GO:0006520")
eggnog_antiox_activity <- eggnog_to_GO_all_annot %>% dplyr::filter(GO_desc == "antioxidant activity")

bact_DESeq_Tw_vs_Vf_l2fc <- read_tsv(file="../data/DESeq2_out/bact_DESeq_Tw_vs_Vf_l2fc.tsv")

bact_DESeq_Tw_vs_Vf_GO_AA_metab <- bact_DESeq_Tw_vs_Vf_l2fc %>% dplyr::filter(feature %in% eggnog_AA_metabolic$OG)

bact_DESeq_Tw_vs_Vf_GO_antiox<-  bact_DESeq_Tw_vs_Vf_l2fc %>% dplyr::filter(feature %in% eggnog_antiox_activity$OG)

Tw_vs_Vf_for_volcano <- site_DE_res_enrichments_GO_MF_GSEA$Tw_vs_Vf$combined_DE_res


#Quick look at labels


Tw_vs_Vf_for_volcano <- merge(Tw_vs_Vf_for_volcano, bact_og_metadata %>% dplyr::rename(feature=bacteria_OG), all.x=TRUE)


###

Tw_vs_Vf_volcano_keyvals_color <- ifelse(
  Tw_vs_Vf_for_volcano$feature %in% intersect(eggnog_antiox_activity$OG %>% unique(.), eggnog_AA_metabolic$OG), 'darkgreen',
      ifelse(Tw_vs_Vf_for_volcano$feature %in% eggnog_antiox_activity$OG, 'gold',
             ifelse(Tw_vs_Vf_for_volcano$feature %in% eggnog_AA_metabolic$OG, 'royalblue',
        'black')))
Tw_vs_Vf_volcano_keyvals_color[is.na(Tw_vs_Vf_volcano_keyvals_color)] <- 'black'
names(Tw_vs_Vf_volcano_keyvals_color)[Tw_vs_Vf_volcano_keyvals_color == 'gold'] <- 'antioxidant activity'
names(Tw_vs_Vf_volcano_keyvals_color)[Tw_vs_Vf_volcano_keyvals_color == 'royalblue'] <- 'AA metabolic process'
names(Tw_vs_Vf_volcano_keyvals_color)[Tw_vs_Vf_volcano_keyvals_color == 'darkgreen'] <- 'Both'


OG_Tw_vs_Vf_DE_volcano <- EnhancedVolcano(Tw_vs_Vf_for_volcano,
                lab = Tw_vs_Vf_for_volcano$feature, 
                title = "Bacteria OGs, Tw vs Vf",
                x='log2FoldChange',
                y='padj',
                #boxedLabels = TRUE,
                drawConnectors = TRUE,
                selectLab = c("COG0599", "COG2128", "COG0753", "COG0625",
                              "COG2986", "COG1228", "COG0112", "COG2159",
                              "COG0407","COG1232","COG0113","COG1587","COG0008","COG0001"),
                labSize = 3.0,
                labFace = 'bold',
                pCutoff = 0.05,
                FCcutoff=0,
                legendPosition = 'top',
              colCustom = Tw_vs_Vf_volcano_keyvals_color
               )


ggsave(filename="../plots/OG_Tw_vs_Vf_DE_volcano.pdf",
       plot=OG_Tw_vs_Vf_DE_volcano,
       height = 8, width = 10)

OG_Tw_vs_Vf_DE_volcano


```




